// tools/tools.js
const window_bar = require('window-bar')

const sheet = new CSSStyleSheet
sheet.replaceSync(get_theme())
/******************************************************************************
  TOOLS COMPONENT
******************************************************************************/
// ----------------------------------------
// MODULE STATE & ID
var count = 0
const [cwd, dir] = [process.cwd(), __filename].map(x => new URL(x, 'file://').href)
const ID = dir.slice(cwd.length)
const STATE = { ids: {}, net: {} } // all state of component module
// ----------------------------------------
const default_opts = { }

module.exports = tools

function tools (opts = default_opts, protocol) {
  // ----------------------------------------
  // ID + JSON STATE
  // ----------------------------------------
  const id = `${ID}:${count++}` // assigns their own name
  const status = {}
  const state = STATE.ids[id] = { id, status, wait: {}, net: {}, aka: {} } // all state of component instance
  const pool = {}
  const name = `tools`
  // ----------------------------------------
  // OPTS
  // ----------------------------------------
  const { data } = opts
  // Assigning all the icons
  const { img_src } = data
  const {
    icon_folder,
    icon_discord,
    icon_github,
  } = img_src
  // ----------------------------------------
  // PROTOCOL
  // ----------------------------------------
  protocol({ from: name }, listen)
  function listen () {
    tools_wrapper.style.display = 'inline'
  }
  // ----------------------------------------
  // TEMPLATE
  // ----------------------------------------
  const el = document.createElement('div')
  const shopts = { mode: 'closed' }
  const shadow = el.attachShadow(shopts)
  shadow.adoptedStyleSheets = [sheet]
  shadow.innerHTML = `<div class="tools">
    <div class="windowbar"></div>
    <div class="tools_content">
      <div class="icon">
        ${icon_discord}
        <span>discord link</span>
      </div>
      <div class="icon">
        ${icon_github}
        <span>github link</span>
      </div>
    </div>
  </div>`
  const tools_wrapper = shadow.querySelector('.tools')
  // ----------------------------------------
  const windowbar_shadow = shadow.querySelector('.windowbar').attachShadow(shopts)
  // ----------------------------------------
  // ELEMENTS
  // ----------------------------------------
  const window = window_bar({
    name:'tools.md', 
    src: icon_folder,
    data: data
  }, tools_protocol)
  windowbar_shadow.append(window)
  // ----------------------------------------
  // INIT
  // ----------------------------------------

  return el

  // cover protocol
  function tools_protocol (message, send) {
    return listen
    // Listening to toggle event 
    function listen (message) {
      const { head, refs, type, data, meta } = message  
      const PROTOCOL = {
        'toggle_active_state': toggle_active_state
      }
      const action = PROTOCOL[type] || invalid      
      action(message)
    }
    function invalid (message) { console.error('invalid type', message) }
    async function toggle_active_state (message) {
      const { head, refs, type, data, meta } = message
      const {active_state} = data
      ;(active_state === 'active') ? tools_wrapper.style.display = 'none' : ''
    }
  }
}

function get_theme () {
  return `
    * {
      box-sizing: border-box;
    }
    .tools {
      display: none;
      .tools_content {
        position: relative;
        display: flex;
        padding: 10px;
        width: 100vw;
        height: 100vh;
        background-size: 10px 10px;
        background-color: var(--bg_color);
        border: 1px solid var(--primary_color);
        gap: 25px;
        margin-bottom: 30px;
        .icon {
          display: flex;
          flex-direction: column;
          gap: 5px;
          align-items: center;
          svg {
            width: 50px;
            height: 50px;
          }
        }
      }
    }
    @container (min-width: 510px) {
      .tools {
        .tools_content {
          width: auto;
          height: auto;
        }
      }
    }
  `
}
