// tools/tools.js
const window_bar = require('window-bar')
const scrollbar = require('scrollbar')

/******************************************************************************
  TOOLS COMPONENT
******************************************************************************/
// ----------------------------------------
// MODULE STATE & ID
var count = 0
const [cwd, dir] = [process.cwd(), __filename].map(x => new URL(x, 'file://').href)
const ID = dir.slice(cwd.length)
const STATE = { ids: {}, net: {} } // all state of component module
// ----------------------------------------
const sheet = new CSSStyleSheet
sheet.replaceSync(get_theme())
const default_opts = { }
const shopts = { mode: 'closed' }
// ----------------------------------------
module.exports = tools
// ----------------------------------------
function tools (opts = default_opts, protocol) {
  // ----------------------------------------
  // RESOURCE POOL (can't be serialized)
  // ----------------------------------------
  const ro = new ResizeObserver(entries => {
    console.log('ResizeObserver:terminal:resize')
    const scroll_channel = state.net[state.aka.scrollbar]
    scroll_channel.send({
      head: [id, scroll_channel.send.id, scroll_channel.mid++],
      refs: { },
      type: 'handle_scroll',
    })
  })
  // ----------------------------------------
  // ID + JSON STATE
  // ----------------------------------------
  const id = `${ID}:${count++}` // assigns their own name
  const status = {}
  const state = STATE.ids[id] = { id, status, wait: {}, net: {}, aka: {} } // all state of component instance
  const cache = resources({})
  // ----------------------------------------
  // OPTS
  // ----------------------------------------
  const { data } = opts
  // Assigning all the icons
  const { img_src } = data
  const {
    icon_folder,
    icon_discord,
    icon_github,
    icon_mastodon,
    icon_opencollective,
    icon_matrix,
    icon_twitter_smooth,
    icon_github_smooth,
    icon_cabal,
    icon_jitsi,
    icon_discord_smooth,
    icon_bigbluebutton,
    icon_youtube,
    icon_hackmd,
    icon_protonmail,
    icon_reddit,
    icon_keet,
  } = img_src
  // ----------------------------------------
  // PROTOCOL
  // ----------------------------------------
  const on = { 'show': on_show, 'hide': on_hide }
  const channel = use_protocol('up')({ protocol, state, on })
  // ----------------------------------------
  // TEMPLATE
  // ----------------------------------------
  const el = document.createElement('div')
  const shadow = el.attachShadow(shopts)
  shadow.adoptedStyleSheets = [sheet]
  shadow.innerHTML = `<div class="tools">
    <div class="windowbar"></div>
    <div class="scrollbar_wrapper">
      <div class="tools_content">
        <a href="https://fosstodon.org/@dat_ecosystem" target="_blank"> 
          <div class="icon"> ${icon_mastodon} Mastodon </div> 
        </a>
        <a href="https://opencollective.com/dat" target="_blank"> 
          <div class="icon"> ${icon_opencollective} OpenCollective </div> 
        </a>
        <a href="https://matrix.to/#/%23datproject_discussions:gitter.im" target="_blank"> 
          <div class="icon"> ${icon_matrix} Matrix </div> 
        </a>
        <a href="https://twitter.com/dat_ecosystem" target="_blank"> 
          <div class="icon"> ${icon_twitter_smooth} Twitter </div> 
        </a>
        <a href="https://github.com/dat-ecosystem" target="_blank"> 
          <div class="icon"> ${icon_github_smooth} github </div> 
        </a>
        <a href="https://github.com/dat-ecosystem/dat-ecosystem.github.io/blob/main/README.md#connect-to-cabal-with-cli-or-download-cabal-desktop" target="_blank"> 
          <div class="icon"> ${icon_cabal} Cabal </div> 
        </a>
        <a href="https://meet.jit.si/dat-ecosystem" target="_blank"> 
          <div class="icon"> ${icon_jitsi} Jitsi </div> 
        </a>
        <a href="https://discord.gg/egsvGc9TkQ" target="_blank"> 
          <div class="icon"> ${icon_discord_smooth} discord </div> 
        </a>
        <a href="https://bigbluebutton.org/" target="_blank"> 
          <div class="icon"> ${icon_bigbluebutton} BigBlueButton </div> 
        </a>
        <a href="https://www.youtube.com/@DatEcosystem-" target="_blank"> 
          <div class="icon"> ${icon_youtube} Youtube </div> 
        </a>
        <a href="https://hackmd.io/@T6Wf5EsOQKe-6wyPjJPtuw/Hycn0F63r/%2Fx_4tQHwtT3u7vrksrposHw" target="_blank"> 
          <div class="icon"> ${icon_hackmd} HackMD </div> 
        </a>
        <a href="dat-ecosystem@protonmail.com" target="_blank"> 
          <div class="icon"> ${icon_protonmail} Protonmail </div> 
        </a>
        <a href="https://www.reddit.com/r/dat_ecosystem/" target="_blank"> 
          <div class="icon"> ${icon_reddit} Reddit </div> 
        </a>
        <a href="https://keet.io/" target="_blank"> 
          <div class="icon"> ${icon_keet} Keet </div> 
        </a>
      </div>
    </div>
  </div>`
  const tools_wrapper = shadow.querySelector('.tools')
  const tools_content = shadow.querySelector('.tools_content')
  const scrollbar_wrapper = shadow.querySelector('.scrollbar_wrapper')
  // ----------------------------------------
  const windowbar_shadow = shadow.querySelector('.windowbar').attachShadow(shopts)
  // ----------------------------------------
  // ELEMENTS
  // ----------------------------------------
  { // windowbar
    const on = {
      'toggle_active_state': toggle_active_state
    }
    const protocol = use_protocol('windowbar')({ state, on })
    const opts = {
      name: 'tools/', 
      src: icon_folder,
      close_button: true,
      data: data
    }
    const element = window_bar(opts, protocol)
    windowbar_shadow.append(element)
    async function toggle_active_state (message) {
      const { active_state } = message.data
      if (active_state === 'active') tools_wrapper.style.display = 'none'
    }
  }
  { // scrollbar
    const on = { 'set_scroll': on_set_scroll, status: onstatus }
    const protocol = use_protocol('scrollbar')({ state, on })
    opts.data.img_src.icon_arrow_start = opts.data.img_src.icon_arrow_up
    opts.data.img_src.icon_arrow_end = opts.data.img_src.icon_arrow_down
    const opts1 = { data }
    const element = shadowfy()(scrollbar(opts1, protocol))
    scrollbar_wrapper.append(element)
    tools_content.onscroll = on_scroll
  }
  // ----------------------------------------
  // INIT
  // ----------------------------------------
  watch_scrollbar()

  return el

  function on_show (event) {
    tools_wrapper.style.display = 'inline'
  }
  function on_hide (event) {
    tools_wrapper.style.display = 'none'
  }
  function watch_scrollbar () {
    ro.observe(tools_content)
  }
  function on_scroll (message) {
    const channel = state.net[state.aka.scrollbar]
    channel.send({
      head: [id, channel.send.id, channel.mid++],
      refs: { },
      type: 'handle_scroll',
    })
  }
  function on_set_scroll (message) {
    console.log('set_scroll', message) 
    setScrollTop(message.data)
  }
  function onstatus (message) {
    const channel = state.net[state.aka.scrollbar]
    channel.send({
      head: [id, channel.send.id, channel.mid++],
      refs: { cause: message.head },
      type: 'update_size',
      data: {
        sh: tools_content.scrollHeight,
        ch: tools_content.clientHeight,
        st: tools_content.scrollTop
      }
    })
  }
  async function setScrollTop (value) {
    tools_content.scrollTop = value
  }
}

function get_theme () {
  return `
    * {
      box-sizing: border-box;
    }
    .tools {
      display: none;
    }
    .tools .scrollbar_wrapper{
      display: flex;
      border: 1px solid var(--primary_color);
      --s: 15px; /* control the size */
      --_g: var(--bg_color_2) /* first color */ 0 25%, #0000 0 50%;
      background:
        repeating-conic-gradient(at 33% 33%,var(--_g)),
        repeating-conic-gradient(at 66% 66%,var(--_g)),
        var(--bg_color_3);  /* second color */  
        background-size: var(--s) var(--s);  
    }
    .tools_content {
      position: relative;
      display: grid;
      grid-template-columns: 4fr 4fr;
      padding: 10px;
      width: calc(100vw - 32px);
      height: calc(100vh - 72px);
      background-size: 10px 10px;
      background-color: var(--bg_color);
      border: 1px solid var(--primary_color);
      gap: 25px;
      overflow-y: scroll;
      scrollbar-width: none; /* For Firefox */
    }
    .tools_content::-webkit-scrollbar{
      display: none;
    }
    .tools_content a{
      text-decoration:none;
    }
    .tools_content .icon {
      display: flex;
      flex-direction: column;
      gap: 5px;
      align-items: center;
      color: var(--primary_color);
    }
    .tools_content .icon svg {
      width: 50px;
      height: 50px;
    }
    path{
      fill: var(--primary_color);
      stroke: var(--primary_color);
    }
    @container (min-width: 510px) {
      .tools_content {
        grid-template-columns: 4fr 4fr 4fr;
        width: auto;
        height: auto;
      }
      .tools{
        margin-bottom: 30px;
      }
    }
  `
}
// ----------------------------------------------------------------------------
function shadowfy (props = {}, sheets = []) {
  return element => {
    const el = Object.assign(document.createElement('div'), { ...props })
    const sh = el.attachShadow(shopts)
    sh.adoptedStyleSheets = sheets
    sh.append(element)
    return el
  }
}
function use_protocol (petname) {
  return ({ protocol, state, on = { } }) => {
    if (petname in state.aka) throw new Error('petname already initialized')
    const { id } = state
    const invalid = on[''] || (message => console.error('invalid type', message))
    if (protocol) return handshake(protocol(Object.assign(listen, { id })))
    else return handshake
    // ----------------------------------------
    // @TODO: how to disconnect channel
    // ----------------------------------------
    function handshake (send) {
      state.aka[petname] = send.id
      const channel = state.net[send.id] = { petname, mid: 0, send, on }
      return protocol ? channel : Object.assign(listen, { id })
    }
    function listen (message) {
      const [from] = message.head
      const by = state.aka[petname]
      if (from !== by) return invalid(message) // @TODO: maybe forward
      console.log(`[${id}]:${petname}>`, message)
      const { on } = state.net[by]
      const action = on[message.type] || invalid
      action(message)
    }
  }
}
// ----------------------------------------------------------------------------
function resources (pool) {
  var num = 0
  return factory => {
    const prefix = num++
    const get = name => {
      const id = prefix + name
      if (pool[id]) return pool[id]
      const type = factory[name]
      return pool[id] = type()
    }
    return Object.assign(get, factory)
  }
}
