// app_projects/index.js
module.exports = app_projects

// Importing components
const project_card = require('project_card')
const window_bar = require('window_bar')
const my_theme = require('my_theme')


// CSS Boiler Plat
const sheet = new CSSStyleSheet
const theme = get_theme()
sheet.replaceSync(theme)

function app_projects(opts, protocol){

     // Assigning all the icons
     const { img_src: {
        icon_discord = `${prefix}/icon_discord.png`,
        icon_twitter = `${prefix}/icon_twitter.png`,
        icon_github = `${prefix}/icon_github.png`,
        icon_folder = `${prefix}/icon_folder.svg`,
        project_logo_1 = `${prefix}/project_logo_1.png`,
    } } = my_theme


    
    const el = document.createElement('div')
    const shadow = el.attachShadow( { mode: 'closed' } )
    shadow.innerHTML = `
        <div class="main_wrapper">
            <div class="project_wrapper"></div>
        </div>
        <style> ${get_theme} </style>
    `

    // Listening to toggle event 
    const listen = (props) =>{
        const {active_state} = props
        ;(active_state==='active')?el.style.display = 'none':''
    }

    // Adding applcation window bar
    const app_project_window = window_bar({
        name: 'OUR_PROJECTS',
        src: icon_folder
    }, listen)

    // Adding project cards
    const project_wrapper = shadow.querySelector('.project_wrapper')
    const cardsData = [
        { 
            title: 'Official starting of the web course.',
            project_logo: project_logo_1,
            project: 'Agregore', 
            link: '/',
            socials: [icon_github, icon_twitter, icon_discord],
            desc: 'Keep track of whānau whakapapa information, preserve and share cultural records and narratives, own and control whānau data and servers, and build a stronger sense of whānau, community and identity.', 
            tags: ['Hypercore', 'Hypercore', 'Hypercore'],
        },{ 
            title: 'Official starting of the web course.',
            project_logo: project_logo_1,
            project: 'Agregore', 
            link: '/',
            socials: [icon_github, icon_twitter, icon_discord],
            desc: 'Keep track of whānau whakapapa information, preserve and share cultural records and narratives, own and control whānau data and servers, and build a stronger sense of whānau, community and identity.', 
            tags: ['Hypercore', 'Hypercore', 'Hypercore'],
        },{ 
            title: 'Official starting of the web course.',
            project_logo: project_logo_1,
            project: 'Agregore', 
            link: '/',
            socials: [icon_github, icon_twitter, icon_discord],
            desc: 'Keep track of whānau whakapapa information, preserve and share cultural records and narratives, own and control whānau data and servers, and build a stronger sense of whānau, community and identity.', 
            tags: ['Hypercore', 'Hypercore', 'Hypercore'],
        },
    ]
    const project_cards = cardsData.map((card_data) => project_card(card_data))
    project_cards.forEach((card) => {
        project_wrapper.append(card)
    })

    shadow.adoptedStyleSheets = [ sheet ]
    shadow.prepend(app_project_window)
    return el

}

function get_theme(){
    return`
        .main_wrapper{
            container-type: inline-size;
            width:100%;
            height: 100%;
        }
        *{
            box-sizing: border-box;
        }
        .project_wrapper{
            --s: 20px; /* control the size */
            --_g: var(--bg_color) /* first color */ 0 25%, #0000 0 50%;
            background:
                repeating-conic-gradient(at 66% 66%,var(--_g)),
                repeating-conic-gradient(at 33% 33%,var(--_g)),
                var(--primary_color);  /* second color */ 
            background-size: var(--s) var(--s);  
            border:1px solid var(--primary_color);
            width:100%;
            height: 100%;
            padding: 0px;
            display: grid;
            gap:20px;
            grid-template-columns: 12fr;
            margin-bottom: 30px;
            box-sizing: border-box;
        }

        @container(min-width: 768px) {
            .project_wrapper{
                grid-template-columns: repeat(2, 6fr);
            }
        }

        @container(min-width: 1200px) {
            .project_wrapper{
                grid-template-columns: repeat(3, 4fr);
            }
        }

        /*---------- Mobile devices ----------*/
        @media (min-width: 480px) {
        }

        /*---------- iPads, Tablets ----------*/
        @media (min-width: 768px) {
        }

        /*---------- Mediuem screens, laptops ----------*/
        @media (min-width: 1024px) {}
    `
}
