// my_theme/index.js
const path = require('path')
const cwd = process.cwd()
const prefix = path.relative(cwd, __dirname)


const my_theme = {
  img_src:{
    icon_consortium: `${prefix}/icon_consortium_page.png`,
    icon_blogger: `${prefix}/icon_blogger.png`,
    icon_discord: `${prefix}/icon_discord.png`,
    icon_twitter: `${prefix}/icon_twitter.png`,
    icon_github: `${prefix}/icon_github.png`,
    icon_terminal: `${prefix}/icon_terminal.png`,
    icon_theme: `${prefix}/icon_theme.png`,
    icon_close_dark: `${prefix}/icon_close_dark.svg`,
    icon_close_light: `${prefix}/icon_close_light.svg`,
    icon_pdf_reader: `${prefix}/icon_pdf_reader.svg`,
    icon_arrow_down: `${prefix}/icon_arrow_down.svg`,
    icon_arrow_up: `${prefix}/icon_arrow_up.svg`,
    icon_arrow_down_light: `${prefix}/icon_arrow_down_light.svg`,
    icon_arrow_up_light: `${prefix}/icon_arrow_up_light.svg`,
    icon_arrow_up_light: `${prefix}/icon_arrow_up_light.svg`,
    banner_cover : `${prefix}/banner_cover.svg`,
    tree_character : `${prefix}/tree_character.png`,
  }
}

module.exports = my_theme
