// timeline-card/timeline-card.js
const sheet = new CSSStyleSheet
sheet.replaceSync(get_theme())
/******************************************************************************
  TIMELINE CARD COMPONENT
******************************************************************************/
// ----------------------------------------
// MODULE STATE & ID
var count = 0
const [cwd, dir] = [process.cwd(), __filename].map(x => new URL(x, 'file://').href)
const ID = dir.slice(cwd.length)
const STATE = { ids: {}, net: {} } // all state of component module
// ----------------------------------------
const default_opts = { }

module.exports = timeline_card

function timeline_card (opts = default_opts) {
  // ----------------------------------------
  // ID + JSON STATE
  // ----------------------------------------
  const name = `timeline_card-${count++}`
  // ----------------------------------------
  // OPTS
  // ----------------------------------------
  const { data, date, time, link, title, desc, tags} = opts
  // Assigning all the icons
  const { img_src } = data
  const {
    icon_clock,
    icon_link,
    icon_calendar,
  } = img_src
  // ----------------------------------------
  // PROTOCOL
  // ----------------------------------------
  const PROTOCOL = {}
  // ----------------------------------------
  // TEMPLATE
  // ----------------------------------------
  const el = document.createElement('div')
  const shopts = { mode: 'closed' }
  const shadow = el.attachShadow(shopts)
  shadow.adoptedStyleSheets = [sheet]
  shadow.innerHTML = `<div class="timeline_card">
    <div class="content_wrapper">
      <div class="icon_wrapper">
        <div> ${icon_calendar} ${date} </div>
        <div> ${icon_clock} ${time} </div>
        <div> <a href="${link}">${icon_link}</a> </div>
      </div>
      <div class="title"> ${title} </div>
      <div class="desc"> ${desc}</div>
    </div>
    <div class="tags_wrapper">
      ${tags.map((tag) => `<div class="tag">${tag}</div>`).join('')}
    </div>
  </div>`
  // ----------------------------------------
  // INIT
  // ----------------------------------------

  return el
}
function get_theme () {
  return `
    * {
      box-sizing: border-box;
    }
    .timeline_card {
      height: max-content;
      width: 100%;
      line-height: normal;
      background-color: var(--bg_color);
      color: var(--primary_color) !important;
      border: 1px solid var(--primary_color);
      container-type: inline-size;
      .content_wrapper {
        padding: 20px;
        .icon_wrapper {
          display: flex;
          gap: 20px;
          div {
            display: flex;
            gap: 5px;
            font-size: 16px;
            letter-spacing: -2px;
            align-items: center;
          }
          img {
            width: 20px;
            height: 20px;
          }
          div:nth-last-child(1) {
            margin-left: auto;
          }
        }
        .title {
          margin-top: 20px;
          margin-bottom: 5px;
          font-size: 18px;
          font-weight: 700;
          letter-spacing: -2px;
          line-height: 16px;
        }
        .desc {
          font-size: 14px;
          letter-spacing: -2px;
          line-height: 16px;
        }
      }
      .tags_wrapper {
        display: flex;
        flex-wrap: wrap;
        .tag {
          flex-grow: 1;
          min-width: max-content;
          padding: 5px 10px;
          border: 1px solid var(--primary_color);
          // line-height:0px;
          text-align: center;
        }
      }
    }
  `
}
