// tab-window/tab-window.js
const scrollbar = require('scrollbar')
/******************************************************************************
  TAB WINDOW COMPONENT
******************************************************************************/
// ----------------------------------------
// MODULE STATE & ID
var count = 0
const [cwd, dir] = [process.cwd(), __filename].map(x => new URL(x, 'file://').href)
const ID = dir.slice(cwd.length)
const STATE = { ids: {}, net: {} } // all state of component module
// ----------------------------------------
const sheet = new CSSStyleSheet
sheet.replaceSync(get_theme())
const default_opts = { }
const shopts = { mode: 'closed' }
// ----------------------------------------
module.exports = tab_window
// ----------------------------------------
function tab_window (opts = default_opts, protocol) {
  // ----------------------------------------
  // RESOURCE POOL (can't be serialized)
  // ----------------------------------------
  const ro = new ResizeObserver(entries => {
    console.log('ResizeObserver:terminal:resize')
    const scroll_channel = state.net[state.aka.scrollbar]
    scroll_channel.send({
      head: [id, scroll_channel.send.id, scroll_channel.mid++],
      refs: { },
      type: 'handle_scroll',
    })
  })
  // ----------------------------------------
  // ID + JSON STATE
  // ----------------------------------------
  const id = `${ID}:${count++}` // assigns their own name
  const status = {}
  const state = STATE.ids[id] = { id, status, wait: {}, net: {}, aka: {} } // all state of component instance
  const cache = resources({})
  status.commands = {
    'help': on_help,
    'list': on_list,
    'goto': on_goto,
    'unknown': 'Command not found'
  }
  // ----------------------------------------
  // OPTS
  // ----------------------------------------
  const { text } = opts
  // ----------------------------------------
  // PROTOCOL
  // ----------------------------------------
  const on = { 
    'toggle_fullscreen': toggle_fullscreen
  }
  const channel = use_protocol('up')({ protocol, state, on })
  // ----------------------------------------
  // TEMPLATE
  // ----------------------------------------
  const el = document.createElement('div')
  const shadow = el.attachShadow(shopts)
  shadow.adoptedStyleSheets = [sheet]
  shadow.innerHTML = `<div class="tab_wrapper">
    <div class="scrollbar_wrapper">
      <div>
        Type help() for more info.
      </div>
      <div class="history">
      </div>
      <div class="dollar">
        <textarea type="text" id="input" autocomplete="off"></textarea>
      </div>
    </div>
  </div>`
  const tab_wrapper = shadow.querySelector('.tab_wrapper')
  const scrollbar_wrapper = shadow.querySelector('.scrollbar_wrapper')
  const history = shadow.querySelector('.history')
  const input = shadow.querySelector('textarea')
  // ----------------------------------------
  // ELEMENTS
  // ----------------------------------------
  input.onkeyup = textAreaAdjust
  input.onkeydown = submitOnEnter
  
  { // scrollbar
    const on = { 'set_scroll': on_set_scroll, 'status': on_update_size }
    const protocol = use_protocol('scrollbar')({ state, on })
    opts.horizontal = false
    opts.data.img_src.icon_arrow_start = opts.data.img_src.icon_arrow_up
    opts.data.img_src.icon_arrow_end = opts.data.img_src.icon_arrow_down
    const scroll_opts = opts
    const element = scrollbar(scroll_opts, protocol)

    scrollbar_wrapper.onscroll = on_scroll
    const scroll_channel = state.net[state.aka.scrollbar]
    ro.observe(scrollbar_wrapper)

    tab_wrapper.append(element)

    function on_scroll (event) {
      console.log('scrollbar_wrapper:terminal:scroll')
      scroll_channel.send({
        head: [id, scroll_channel.send.id, scroll_channel.mid++],
        refs: { },
        type: 'handle_scroll',
      })
    }
    function on_set_scroll (message) {
      console.log('set_scroll', message) 
      setScrollTop(message.data)
    }
    function on_update_size (message) {
      const head = [id, scroll_channel.send.id, scroll_channel.mid++]
      scroll_channel.send({
        head,
        refs: { cause: message.head },
        type: 'update_size',
        data: {
          sh: scrollbar_wrapper.scrollHeight,
          ch: scrollbar_wrapper.clientHeight,
          st: scrollbar_wrapper.scrollTop
        }
      })
    }
    async function setScrollTop (value) {
      scrollbar_wrapper.scrollTop = value
    }
  }
  // ----------------------------------------
  // INIT
  // ----------------------------------------

  return el

  function textAreaAdjust (e) {
    e.target.style.height = "1px";
    e.target.style.height = (25+e.target.scrollHeight)+"px";
  }
  function submitOnEnter (e) {
    if (e.which === 13){
      if (input.value) {
        const prompt = document.createElement('div')
        prompt.innerHTML = `$ ${input.value}`
        
        const [command, data] = input.value.toLowerCase().split('(')
        
        const response = document.createElement('div')
        if(status.commands[command])
          response.innerHTML = status.commands[command](data.replace(/[)'"']/g, ''))
        else
          response.innerHTML = status.commands['unknown']

        history.append(...[prompt, response])
      }
      else{
        const prompt = document.createElement('div')
        prompt.innerHTML = '$'
        history.append(prompt)
      }
      //Clear input
      setTimeout(() => input.value = '', 2)
    }
  }
  function on_list (data){
    const command_list = {
      info: {
        'mission_statement.md': 'mission_statement',
        'important_documents.md': 'important_documents',
        'our_members.md': 'our_members',
        'tools': {}
      },
      home: {},
      projects: {},
      timeline: {}
    }
    const response = data ? command_list[data] : command_list
    if(typeof(response) === 'object'){
      return '[' + Object.entries(response).map(v => {
        if(typeof(v[1]) === 'string')
          return v[0]
        else
          return v[0] + '/'
      }).toString().replace(/,/g, ', ') + ']'
    }
    else{
      return response
    }
  }
  function on_help (){
    const command_list = { 
      help: 'description of help', 
      list: 'description of list command', 
      goto: '...', 
      read: '...' 
    }

    return JSON.stringify(command_list, null, 2).replace(/"/g, '')
  }
  function on_goto (data){
    const up_channel = state.net[state.aka.up]
    up_channel.send({
      head: [id, up_channel.send.id, up_channel.mid++],
      type: 'navigate',
      data: data.slice(0, -1).toUpperCase()
    })
    return ''
  }
  function toggle_fullscreen (msg){
    scrollbar_wrapper.classList.toggle('fullscreen')
  }

}

function get_theme() {
  return `
    .tab_wrapper{
      display: flex;
    }
    .tab_wrapper > div:last-child{
      border-left: 1px solid var(--primary_color);
    }
    .scrollbar_wrapper{
      max-height: 220px;
      overflow-y: scroll;
      overflow-x: clip;
      width: 100%;
    }
    .scrollbar_wrapper.fullscreen{
      max-height: calc(100vh - 80px);
    }
    .scrollbar_wrapper::-webkit-scrollbar {
      display: none;
    }
    .tab_wrapper #input{
      display: inline;
      background-color: transparent;
      color: var(--primary-color);
      border: none;
      outline: none;
      width: 100%;
      text-indent: 20px;
      font-family: Silkscreen;
      font-size: 16px;
      padding: 0;
      resize: none;
    }
    .tab_wrapper .dollar{
      position: relative;
    }
    .tab_wrapper .dollar::before{
      content: '$';
      position: absolute;
      top: 0;
    }
  `
}
// ----------------------------------------------------------------------------
function shadowfy (props = {}, sheets = []) {
  return element => {
    const el = Object.assign(document.createElement('div'), { ...props })
    const sh = el.attachShadow(shopts)
    sh.adoptedStyleSheets = sheets
    sh.append(element)
    return el
  }
}
function use_protocol (petname) {
  return ({ protocol, state, on = { } }) => {
    if (petname in state.aka) throw new Error('petname already initialized')
    const { id } = state
    const invalid = on[''] || (message => console.error('invalid type', message))
    if (protocol) return handshake(protocol(Object.assign(listen, { id })))
    else return handshake
    // ----------------------------------------
    // @TODO: how to disconnect channel
    // ----------------------------------------
    function handshake (send) {
      state.aka[petname] = send.id
      const channel = state.net[send.id] = { petname, mid: 0, send, on }
      return protocol ? channel : Object.assign(listen, { id })
    }
    function listen (message) {
      const [from] = message.head
      const by = state.aka[petname]
      if (from !== by) return invalid(message) // @TODO: maybe forward
      console.log(`[${id}]:${petname}>`, message)
      const { on } = state.net[by]
      const action = on[message.type] || invalid
      action(message)
    }
  }
}
// ----------------------------------------------------------------------------
function resources (pool) {
  var num = 0
  return factory => {
    const prefix = num++
    const get = name => {
      const id = prefix + name
      if (pool[id]) return pool[id]
      const type = factory[name]
      return pool[id] = type()
    }
    return Object.assign(get, factory)
  }
}
