// app-about-us/app-about-us.js
const window_bar = require('window-bar')

const sheet = new CSSStyleSheet
sheet.replaceSync(get_theme())
/******************************************************************************
  WINDOW BAR COMPONENT
******************************************************************************/
// ----------------------------------------
// MODULE STATE & ID
var count = 0
const [cwd, dir] = [process.cwd(), __filename].map(x => new URL(x, 'file://').href)
const ID = dir.slice(cwd.length)
const STATE = { ids: {}, net: {} } // all state of component module
// ----------------------------------------
const default_opts = { }

module.exports = app_about_us

function app_about_us (opts = default_opts) {
  // ----------------------------------------
  // ID + JSON STATE
  // ----------------------------------------
  const id = count++
  // ----------------------------------------
  // OPTS
  // ----------------------------------------
  const { data } = opts
  // Assigning all the icons
  const { img_src: { 
    about_us_cover = `${prefix}/about_us_cover.png`,
    img_robot_1 = `${prefix}/img_robot_1.svg`,
    icon_pdf_reader = `${prefix}/icon_pdf_reader.svg`,
  } } = data
  // ----------------------------------------
  // PROTOCOL
  // ----------------------------------------
  const PROTOCOL = {}
  // ----------------------------------------
  // TEMPLATE
  // ----------------------------------------
  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode : 'closed' })
  shadow.innerHTML = `<div class="about_us_wrapper">
    <div class="about_us_cover_image">
      <img src="${about_us_cover}"/>
    </div>
    <div class="content_wrapper">
      <img src="${img_robot_1}"/>
      <div class="title"> ABOUT US </div>
    </div>
  </div>
  <div class="about_us_desc">
    Dat ecosystem garden supports open source projects that strengthen P2P foundations, with a focus on builder tools, infrastructure, research, and community resources.
  </div>`
  shadow.adoptedStyleSheets = [sheet]
  // ----------------------------------------
  // ELEMENTS
  // ----------------------------------------
  const cover_window = window_bar({
    name:'Learn_about_us.pdf', 
    src: icon_pdf_reader,
    action_buttons: ['IMPORTANT DOCUMENTS', 'TELL ME MORE'],
    data
  }, about_us_protocol)
  shadow.prepend(cover_window)
  // ----------------------------------------
  // INIT
  // ----------------------------------------

  return el

  // about us protocol
  function about_us_protocol (message, send) {
    return listen
  }
  // Listening to toggle event 
  function listen (message) {
    const { head, refs, type, data, meta } = message  
    const PROTOCOL = {
      'toggle_active_state': toggle_active_state
    }
    const action = PROTOCOL[type] || invalid      
    action(message)
  }
  function invalid (message) { console.error('invalid type', message) }
  async function toggle_active_state (message) {
    const { head, refs, type, data, meta } = message
    const { active_state } = data
    ;(active_state === 'active')?el.style.display = 'none':''
  }
}
function get_theme () {
  return `
    * {
      box-sizing: border-box;
    }
    .about_us_wrapper {
      position:r elative;
      height: max-content;
      width: 100%;
      display: flex;
      flex-direction: column;
      justify-content: center;
      align-items: center;
      padding: 150px 0px;
      background-image: radial-gradient(var(--primary_color) 1px, var(--bg_color) 1px);
      background-size: 10px 10px;
      background-color: red;
      border: 1px solid var(--primary_color);
      box-sizing: border-box;
      container-type: inline-size;
      /* This covers background-image will change to an image */
      .about_us_cover_image {
        position: absolute;
        width: 100%;
        height: 100%;
        overflow: hidden;
        img {
          position: absolute;
          left: 50%;
          top: 50%;
          width: auto;
          height: 80%;
          transform: translate(-50%, -50%);
        }
      }
      /* Cover image alignment */
      .content_wrapper {
        display: flex;
        justify-content: center;
        align-items: center;
        gap: 20px;
        position: relative;
        z-index: 1;
        color: var(--primary_color);
        text-align: center;
        img {
          width: 100px;
          height: auto;
        }
        .title{ 
          font-size: 40px;
        }
      }
    }
    .about_us_desc {
      width: 100% !important;
      background-color: var(--bg_color);
      color: var(--primary_color);
      border: 1px solid var(--primary_color);
      padding: 10px;
      letter-spacing: -2px;
      line-height: 18px;
      font-size: 16px;
      margin-bottom: 30px;
      box-sizing: border-box;
    }
    @container (min-width: 856px) {
      .about_us_cover_image {
        img {
          width: 100%;
          height: auto;
        }
      }
    }
  `
}
