// search_input/index.js
module.exports = input_search

const path = require('path')
const cwd = process.cwd()   
const prefix = path.relative(cwd, __dirname)

// CSS Boiler Plat
const sheet = new CSSStyleSheet
const theme = get_theme()
sheet.replaceSync(theme)


function input_search (opts, protocol){
    const notify = protocol(null, listen)
    const {data} = opts
    let message = {
        head: ['input_search', 'project_filter', 'project_filter'],
        type: 'search',
    }

    // Assigning all the icons
    const { img_src: {
        icon_search = `${prefix}/icon_search.svg`,
    } } = data

    const el = document.createElement('div')
    // el.classList.add('input_wrapper')

    const shadow = el.attachShadow( { mode:`closed` } )
    shadow.innerHTML = `
        <div class="search_input">
            <input class="input" type="text" placeholder="SEARCH...">
                ${icon_search}
            </input>
        </div>
        <style> ${get_theme} </style>
    `
    const input = shadow.querySelector('.input')
    input.onchange = (e) => {
        message['data'] = e.target.value
        notify(message)
    }

    // shadow.append(main, navbar(opts, protocol))
    shadow.adoptedStyleSheets = [sheet]
    return el

    function listen(message){
        // const {head, type, data} = message
        // const [by, to, id] = head
        // if( to !== id) return console.error('address unknown', message)

    }
}


function get_theme(){

    return`
        .search_input{
            width: 100%;
            min-width: 100% !important;
            height: 40px;
            max-height: 40px;
            position: relative;
            flex-grow: 1;
        }
        input{
            box-sizing: border-box;
            width:100%;
            height: 100%;
            border: 3px solid var(--primary_color);
            padding: 10px 40px 10px 5px;
            outline:none;
            font-family: Silkscreen;
            font-size: 18px;
            letter-spacing:-1px;
        }
        input:focus{
            border-color: var(--ac-1) !important;
        }
        .search_input svg{
            position: absolute;
            right:10px;
            top:50%;
            translate: 0 -50%;
            width:20px;
            height: auto;
        }
    `
    
}
