// mission-statement/mission-statement.js
const window_bar = require('window-bar')

const sheet = new CSSStyleSheet
sheet.replaceSync(get_theme())
/******************************************************************************
  MISSION STATEMENT COMPONENT
******************************************************************************/
// ----------------------------------------
// MODULE STATE & ID
var count = 0
const [cwd, dir] = [process.cwd(), __filename].map(x => new URL(x, 'file://').href)
const ID = dir.slice(cwd.length)
const STATE = { ids: {}, net: {} } // all state of component module
// ----------------------------------------
const default_opts = { }

module.exports = mission_statement

function mission_statement (opts = default_opts, protocol) {
  // ----------------------------------------
  // ID + JSON STATE
  // ----------------------------------------
  const name = `mission_statement`
  // ----------------------------------------
  // OPTS
  // ----------------------------------------
  const { data } = opts
  // Assigning all the icons
  const { img_src } = data
  const {
    icon_pdf_reader
  } = img_src
  // ----------------------------------------
  // PROTOCOL
  // ----------------------------------------
  protocol({ from: name }, listen)
  function listen () {
    mission_statement_wrapper.style.display = 'inline'
  }
  // ----------------------------------------
  // TEMPLATE
  // ----------------------------------------
  const el = document.createElement('div')
  const shopts = { mode: 'closed' }
  const shadow = el.attachShadow(shopts)
  shadow.adoptedStyleSheets = [sheet]
  shadow.innerHTML = `<div class="mission_statement">
    <div class="window_bar_wrapper"></div>
    <div class="mission_content">
      <h2>OUR MISSION</h2>
      <p>We aim to connect and support the dat community, promoting user rights and decentralized democracy, dat ecosystem provides resources to advance your hyprecore project.</p>
      <h2>OUR MISSION</h2>
      <p>We aim to connect and support the dat community, promoting user rights and decentralized democracy, dat ecosystem provides resources to advance your hyprecore project.</p>    
    </div>
  </div>`
  const mission_statement_wrapper = shadow.querySelector('.mission_statement')
  // ----------------------------------------
  const window_bar_wrapper = shadow.querySelector('.window_bar_wrapper').attachShadow(shopts)
  // ----------------------------------------
  // ELEMENTS
  // ----------------------------------------
  const window = window_bar({
    name: 'Mission_statement.md', 
    src: icon_pdf_reader,
    data: data
  }, mission_statement_protocol)
  window_bar_wrapper.append(window)
  // ----------------------------------------
  // INIT
  // ----------------------------------------

  return el

  // cover protocol
  function mission_statement_protocol (message, send) {
    return listen
    // Listening to toggle event 
    function listen (message) {
      const { head, refs, type, data, meta } = message  
      const PROTOCOL = {
        'toggle_active_state': toggle_active_state
      }
      const action = PROTOCOL[type] || invalid      
      action(message)
    }
    function invalid (message) { console.error('invalid type', message) }
    async function toggle_active_state (message) {
      const { head, refs, type, data, meta } = message
      const { active_state } = data
      ;( active_state === 'active')?mission_statement_wrapper.style.display = 'none':''
    }
  }
}
function get_theme () {
  return `
    * {
      box-sizing: border-box;
      color: var(--primary_color);
    }
    .mission_statement {
      display: none;
      .mission_content {
        position: relative;
        display: flex;
        flex-direction: column;
        width: 100vw;
        height: 100vh;
        padding: 10px;
        background-size: 10px 10px;
        background-color: var(--bg_color);
        border: 1px solid var(--primary_color);
        margin-bottom: 30px;
        h2 {
          margin: 0;
        }
      }
    } 
    @container (min-width: 510px) {
      .mission_statement {
        .mission_content {
          width: auto;
          height: auto;
        }
      }
    }
  `
}
