// app-footer/app-footer.js
const window_bar = require('window-bar')
const sm_text_button = require('buttons/sm-text-button')

const sheet = new CSSStyleSheet
sheet.replaceSync(get_theme())
/******************************************************************************
  APP FOOTER COMPONENT
******************************************************************************/
// ----------------------------------------
// MODULE STATE & ID
var count = 0
const [cwd, dir] = [process.cwd(), __filename].map(x => new URL(x, 'file://').href)
const ID = dir.slice(cwd.length)
const STATE = { ids: {}, net: {} } // all state of component module
// ----------------------------------------
const default_opts = { }

module.exports = app_footer

function app_footer (opts = default_opts) {
  // ----------------------------------------
  // ID + JSON STATE
  // ----------------------------------------
  const id = count++
  // ----------------------------------------
  // OPTS
  // ----------------------------------------
  const { data } = opts
  // Assigning all the icons
  const { img_src: {
    icon_pdf_reader = `${prefix}/icon_pdf_reader.svg`,
    img_robot_2 = `${prefix}/img_robot_2.png`,
    pattern_img_1 = `${prefix}/pattern_img_1.png`,
  } } = data
  // ----------------------------------------
  // PROTOCOL
  // ----------------------------------------
  const PROTOCOL = {}
  // ----------------------------------------
  // TEMPLATE
  // ----------------------------------------
  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode : 'closed' })
  shadow.innerHTML = `<div class="main_wrapper">
    <div class="footer_wrapper">
      <div class="robot_img_2"><img src="${img_robot_2}"></div>
      <div class="footer_info_wrapper">
        <div class="title"> INTERESTED IN JOINING DAT ECOSYSTEM CHAT NETWORKING? </div>
        <div class="desc"> Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vitae porta aliquet sit amet ornare sagittis, ultricies sed. Viverra sit felis ullamcorper pharetra mattis amet, vel. </div>
        <apply_button></apply_button>    
      </div>
    </div>
    <div class="pattern_img"><img src="${pattern_img_1}"></div>
  </div>`
  shadow.adoptedStyleSheets = [sheet]
  const apply_button = shadow.querySelector('apply_button')
  // ----------------------------------------
  // ELEMENTS
  // ----------------------------------------
  // the following is the pattern we usually use, but what you do is more or less the same
  // so you can also keep your three liner :-)
  const join_programe = sm_text_button({ text: 'JOIN OUR GROWTH PROGRAME' })
  apply_button.replaceWith(join_programe)
  // Adding Footer Window
  const footer_window = window_bar({
    name:'FOOTER.pdf', 
    src: icon_pdf_reader,
    data,
  }, footer_protocol)
  shadow.prepend(footer_window)
  // ----------------------------------------
  // INIT
  // ----------------------------------------

  return el

  // footer protocol
  function footer_protocol (message, send) {
    return listen
  }
  // Listening to toggle event 
  function listen (message) {
    const { head, refs, type, data, meta } = message  
    const PROTOCOL = {
      'toggle_active_state': toggle_active_state
    }
    const action = PROTOCOL[type] || invalid      
    action(message)
  }
  function invalid (message) { console.error('invalid type', message) }
  async function toggle_active_state (message) {
    const { head, refs, type, data, meta } = message
    const { active_state } = data
    ;( active_state === 'active')?el.style.display = 'none':''
  }
}
function get_theme () {
  return `
    * { box-sizing: border-box; }
    .main_wrapper {
      position: relative;
      container-type: inline-size;
      background-color: var(--bg_color);
      border: 1px solid var(--primary_color);
      margin-bottom: 30px;
      .footer_wrapper {
        display: flex;
        flex-direction: column-reverse;
        align-items: flex-start;
        padding: 20px;
        padding-bottom: 0px !important;
        .robot_img_2 img {
          width: 150px;
        }
        .footer_info_wrappe r{
          margin-bottom: 30px;
          .title {
            font-size: 40px;
            color: var(--primary_color);
            font-weight: 700;
            line-height: 36px;
            letter-spacing: -5px;
            margin-bottom: 10px;
          }
          .desc {
            font-size: 16px;
            color: var(--primary_color);
            line-height: 14px;
            letter-spacing: -2px;
            margin-bottom: 30px;
          }
        }
      }
      .pattern_img {
        display:none;
      }
    }
    @container (min-width: 856px) {
      .main_wrapper {
        .footer_wrapper {
          gap: 40px;
          flex-direction: row;
          align-items: flex-end;
          width: 70%;
        }
        .pattern_img {
          display: block;
          position: absolute;
          top: 0;
          right: 0;
          & img {
            width: 300px;
            height: auto;
          }
        }
      }
    }
  `
}
