// project_filter/index.js
module.exports = project_filter


const search_input = require('search_input')
const select_button = require('../buttons/select_button')

// CSS Boiler Plat
const sheet = new CSSStyleSheet
const theme = get_theme()
sheet.replaceSync(theme)

var id = 0;
function project_filter(opts, protocol){

    const name = 'project_filter-'+id++;
    const notify = protocol({from: name}, listen)

    const PROTOCOL = {

    }

    const el = document.createElement('div')
    const shadow = el.attachShadow( { mode:`closed` } )
    shadow.innerHTML = `
        <div class="filter_wrapper">
            <div class="project_filter"></div>
        </div>
        <style> ${get_theme} </style>
    `

    const search_project = search_input(opts, project_filter_protocol)
    const status_button = select_button(opts, project_filter_protocol)
    const tag_button = select_button(opts, project_filter_protocol)

    const project_filter = shadow.querySelector('.project_filter')
    project_filter.append(status_button, tag_button, search_project)

    // shadow.append(project_filter)
    shadow.adoptedStyleSheets = [sheet]
    return el

    function project_filter_protocol(handshake, send, mid = 0){
        if(send) return listen
        function listen(message){
            const {head, type, data} = message
            const {by, to, id} = head
            // if( to !== id) return console.error('address unknown', message)
            message = {
                head: {by:name, to:'app_projects', mid:0},
                type: type,
                data: data
            }
            notify(message)
        }
    }
    function listen(message){
        
    }

}

function get_theme(){
    return`
        .filter_wrapper{
            container-type: inline-size;
        }
        .project_filter{
            display: grid;
            grid-template-columns: 12fr;
            align-items:flex-end;
        }
        .project_filter > div{
            flex-grow: 1;
        }
        .project_filter .select_button_wrapper{
            width: 100% !important;
        }





        @container(min-width: 412px) {
            .project_filter{
                grid-template-columns: 1fr 1fr 10fr;
            }
        }
    `
}
