// our_member/index.js
module.exports = our_member

const window_bar = require('window_bar')

// CSS Boiler Plat
const sheet = new CSSStyleSheet
const theme = get_theme()
sheet.replaceSync(theme)


let id = 0


function our_member (opts, protocol) {
    const name = `our_member`

    protocol({from: name}, listen)
    function listen(){
        our_member_wrapper.style.display = 'inline'
    }

    const {data} = opts
    // Assigning all the icons
    const {img_src} = data
    const {
        icon_pdf_reader
    } = img_src

    const el = document.createElement('div')
    const shadow = el.attachShadow ( { mode : 'closed' } )

    shadow.innerHTML = `
        <div class="our_member">
            <div class="member_content">
                <h2>## our members</h2>
                <table>
                    <thead>
                        <tr>
                            <td>
                                s.no
                            </td>
                            <td>
                                names
                            </td>
                            <td>
                                socials
                            </td>
                        </tr>
                    </thead>
                    <tbody>
                        <tr>
                            <td>
                                01
                            </td>
                            <td>
                                alexander
                            </td>
                            <td>
                                cabal
                            </td>
                        </tr>
                        <tr>
                            <td>
                                02
                            </td>
                            <td>
                                alexander praetorius
                            </td>
                            <td>
                                geut/she
                            </td>
                        </tr>
                    </tbody>
                </table>  
            </div>
        </div>
        <style> ${get_theme()} </style>
    `


    const window = window_bar({
        name:'our_member.md', 
        src: icon_pdf_reader,
        data: data
    }, our_member_protocol)
    const our_member_wrapper = shadow.querySelector('.our_member')
    our_member_wrapper.prepend(window)


    shadow.adoptedStyleSheets = [ sheet ]
    return el

    
    // cover protocol
    function our_member_protocol(message, send){
        return listen
        // Listening to toggle event 
        function listen (message) {
            const {head, refs, type, data, meta} = message  
            const PROTOCOL = {
                'toggle_active_state': toggle_active_state
            }
            const action = PROTOCOL[type] || invalid      
            action(message)
        }
        function invalid (message) { console.error('invalid type', message) }
        async function toggle_active_state (message) {
            const {head, refs, type, data, meta} = message
            const {active_state} = data
            ;( active_state === 'active')?our_member_wrapper.style.display = 'none':''
        }
    }
    

}

function get_theme(){
    return`
        *{
            box-sizing: border-box;
        }
        .our_member{
            display: none;
        }
        .member_content{
            position:relative;
            display:flex;
            flex-direction: column;
            width: 100vw;
            height: 100vh;
            padding: 10px;
            background-size: 10px 10px;
            background-color:var(--bg_color);
            border: 1px solid var(--primary_color);
            margin-bottom: 30px;
        }
        h2{
            margin: 0;
        }
        table{
            border-collapse: collapse;
        }
        td{
            border: 1px solid var(--primary_color);
            padding: 8px;
        }
        thead{
            font-weight: bold;
        }
        @media (min-width: 510px){
            .member_content{
                width: auto;
                height: auto;
            }
        }

    `
}
