// app_cover/index.js
module.exports = cover_app


const path = require('path')
const cwd = process.cwd()
const prefix = path.relative(cwd, __dirname)

const window_bar = require('window_bar')
const my_theme = require('my_theme')
const sm_text_button = require('buttons/sm_text_button')

// CSS Boiler Plat
const sheet = new CSSStyleSheet
const theme = get_theme()
sheet.replaceSync(theme)





function cover_app () {

    // Assigning all the icons
    const { img_src: { 
        banner_cover = `${prefix}/banner_cover.svg`,
        tree_character = `${prefix}/tree_character.png`,
        icon_pdf_reader = `${prefix}/icon_pdf_reader.svg`,
    } } = my_theme;

    const el = document.createElement('div')
    const shadow = el.attachShadow ( { mode : 'closed' } )

    shadow.innerHTML = `
        <div class="cover_wrapper">
            <div class="cover_image"></div>
            <div class="content_wrapper">
                WELCOME TO DAT ECOSYSTEM
            </div>
        </div>
        <style> ${get_theme} </style>
    `

    // Listening to toggle event 
    const listen = (props) =>{
        const {active_state} = props
        ;(active_state==='active')?el.style.display = 'none':''
    }

    // Added background banner cover
    const cover_image = shadow.querySelector('.cover_image')
    banner_img = document.createElement('img')
    banner_img.src = banner_cover
    cover_image.append(banner_img)

    // cover_wrapper.style.backgroundImage = `url(${banner_cover})`
    const content_wrapper = shadow.querySelector('.content_wrapper')
    tree_character_img = document.createElement('img')
    tree_character_img.src = tree_character
    content_wrapper.prepend(tree_character_img)

    const view_more_btn = sm_text_button({text:'View more (20)'})
    const tell_me_more_btn = sm_text_button({text:'TELL ME MORE'})
    const cover_window = window_bar({
        name:'Learn_about_us.pdf', 
        src: icon_pdf_reader,
        action_buttons: [tell_me_more_btn, view_more_btn]
    }, listen)

    shadow.adoptedStyleSheets = [ sheet ]
    shadow.prepend(cover_window)
    return el

}




function get_theme(){
    return`
        :host{ 
            --white: white;
            --ac-1: #2ACA4B;
            --ac-2: #F9A5E4;
            --ac-3: #88559D;
            --ac-4: #293648;
        }

        .app_cover{
            display:none;
        }

        .cover_wrapper{
            position:relative;
            height:max-content;
            width:100%;
            display:flex;
            justify-content: center;
            align-items: center;
            padding: 150px 0px;
            background-image: radial-gradient(#A7A6A4 1px, #FFF 1px);
            background-size: 10px 10px;
            background-color:red;
            border: 1px solid var(--ac-4);
            margin-bottom: 30px;
        }

        /* This covers background-image will change to an image */
        .cover_image{
            position: absolute;
            width:100%;
            height:100%;
            overflow:hidden;
        }
        .cover_image img{
            position:absolute;
            left:50%;
            top:50%;
            width: auto;
            height: 100%;
            transform:translate(-50%, -50%);
        }


        /* Cover image alignment */
        .content_wrapper{
            display: flex;
            flex-direction: column;
            align-items:center;
            gap:20px;
            position: relative;
            z-index:1;
            color:var(--ac-4);
            text-align:center;
        }
        .content_wrapper img{
            width: 300px;
            height: auto;
        }

    `
}
