// window_bar/index.js
module.exports = window_bar


const path = require('path')
const cwd = process.cwd()
const prefix = path.relative(cwd, __dirname)

const icon_button = require('buttons/icon_button')
const text_button = require('buttons/text_button')
const my_theme = require('my_theme')


// CSS Boiler Plat
const sheet = new CSSStyleSheet
const theme = get_theme()
sheet.replaceSync(theme)





function window_bar (props, notify) {

    // Assigning all the icons
    const { img_src: { 
        icon_close_dark = `${prefix}/icon_close_dark.svg`,
        icon_pdf_reader = `${prefix}/icon_pdf_reader.svg`,
        icon_arrow_down = `${prefix}/icon_arrow_down.svg`,
        icon_arrow_up = `${prefix}/icon_arrow_up.svg`
    } } = my_theme;


    const el = document.createElement('div')
    el.style.lineHeight = '0px'
    const shadow = el.attachShadow( { mode : 'closed' } )
    
    shadow.innerHTML = `
        <div class="window_bar">
            <div class="application_icon_wrapper"></div>
            <div class="application_name"><span>${props.name}</span></div>
            <div class="window_bar_actions">
                <div class="actions_wrapper"></div>
            </div>
        </div>
        <style>${get_theme}</style>
    `

    // adding application icon
    const application_icon = icon_button({src:icon_pdf_reader})
    const application_icon_wrapper = shadow.querySelector('.application_icon_wrapper')
    application_icon_wrapper.append(application_icon)

    // adding close window button
    const window_bar_actions = shadow.querySelector('.window_bar_actions')
    const close_window_btn = icon_button({src:icon_close_dark})
    close_window_btn.addEventListener('click', function() {
        notify( { active_state : 'active' } )
    });
    

    // adding additional actions wrapper
    const actions_wrapper = shadow.querySelector('.actions_wrapper')
    const view_more_btn = text_button({text:'View more (20)'})
    const tell_me_more_btn = text_button({text:'TELL ME MORE'})
    actions_wrapper.append(tell_me_more_btn, view_more_btn)

    // adding toggle button for action wrapper
    const actions_toggle_btn = icon_button({ src: icon_arrow_down, src_active: icon_arrow_up })
    actions_toggle_btn.classList.add('actions_toggle_btn')
    actions_toggle_btn.addEventListener('click', function() {
        shadow.querySelector('.window_bar_actions').classList.toggle('active');
    });


    window_bar_actions.append(actions_toggle_btn, close_window_btn)


    shadow.adoptedStyleSheets = [sheet]
    return el


}





function get_theme(){
    return`
        :host{ 
            --white: white;
            --ac-1: #2ACA4B;
            --ac-2: #F9A5E4;
            --ac-3: #88559D;
            --ac-4: #293648;
        }

        .window_bar{
            position: relative;
            z-index:2;
            height:40px;
            background-color: var(--ac-4);
            display:inline-flex;
            width:100%;
            justify-content: flex-start;
            background-size: 10px 10px;
            background-image:  repeating-linear-gradient(0deg, #FFFFFF, #FFFFFF 2px, #293648 2px, #293648);
            container-type: inline-size;
            border: 1px solid var(--ac-4);
        }

        .application_icon_wrapper{ 
            display:none;
        }

        .application_name{
            display:flex;
            align-items:center;
            min-height: 100%;
            width: max-content;
            color:var(--white);
            padding: 0 10px;
            box-sizing:border-box;
            border: 1px solid var(--ac-4);
            background-color:var(--ac-4);
        }

        .window_bar_actions{
            margin-left: auto;
            display:flex;
        }
        .window_bar_actions.active .actions_wrapper{
            display:block;
        }
        .actions_wrapper{
            display:none;
            position: absolute;
            z-index:10;
            width: 100%;
            height:max-content;
            top:40px;
            right:0;
            background-color:var(--white);
            // background-color:red;
            border: 1px solid var(--ac-4);
        }




        @container(min-width: 856px) {
            .application_icon_wrapper{ 
                display:block;
            }

            .actions_toggle_btn{
                display:none;
            }

            .actions_wrapper{
                display: flex;
                position: relative;
                top: unset;
                right: unset;
                height:100%;
                width:max-content;
                border: 0px;
            }
        }


    `
}
