// app_footer/index.js
module.exports = app_footer


const path = require('path')
const cwd = process.cwd()
const prefix = path.relative(cwd, __dirname)

const window_bar = require('window_bar')
const my_theme = require('my_theme')
const sm_text_button = require('buttons/sm_text_button')

// CSS Boiler Plat
const sheet = new CSSStyleSheet
const theme = get_theme()
sheet.replaceSync(theme)





function app_footer () {

    // Assigning all the icons
    const { img_src: {
        icon_pdf_reader = `${prefix}/icon_pdf_reader.svg`,
        img_robot_2 = `${prefix}/img_robot_2.png`,
        pattern_img_1 = `${prefix}/pattern_img_1.png`,
    } } = my_theme;

    const el = document.createElement('div')
    const shadow = el.attachShadow ( { mode : 'closed' } )

    shadow.innerHTML = `
        <div class="main_wrapper">
            <div class="footer_wrapper">
                <div class="robot_img_2"></div>
                <div class="footer_info_wrapper">
                    <div class="title"> INTERESTED IN JOINING DAT ECOSYSTEM CHAT NETWORKING? </div>
                    <div class="desc"> Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vitae porta aliquet sit amet ornare sagittis, ultricies sed. Viverra sit felis ullamcorper pharetra mattis amet, vel. </div>
                </div>
            </div>
            <div class="pattern_img"></div>
        </div>
        <style> ${get_theme} </style>
    `

    // Listening to toggle event 
    const listen = (props) =>{
        const {active_state} = props
        ;(active_state==='active')?el.style.display = 'none':''
    }
    
    // Adding Robot Image
    const robot_image_wrapper = shadow.querySelector('.robot_img_2')
    const robot_image = document.createElement('img')
    robot_image.src = img_robot_2
    robot_image_wrapper.append(robot_image)

    // Adding Button
    const footer_info_wrapper = shadow.querySelector('.footer_info_wrapper')
    const join_programe = sm_text_button({text:'JOIN OUR GROWTH PROGRAME'})
    footer_info_wrapper.append(join_programe)

    // Adding Pattern Image
    const pattern_image_wrapper = shadow.querySelector('.pattern_img')
    const pattern_image = document.createElement('img')
    pattern_image.src = pattern_img_1
    pattern_image_wrapper.append(pattern_image)


    // Adding Footer Window
    const footer_window = window_bar({
        name:'FOOTER.pdf', 
        src: icon_pdf_reader,
    }, listen)

    shadow.adoptedStyleSheets = [ sheet ]
    shadow.prepend(footer_window)
    return el

}




function get_theme(){
    return`
        *{ box-sizing: border-box; }

        .main_wrapper{
            position: relative;
            container-type: inline-size;
            background-color: var(--bg_color);
            border: 1px solid var(--primary_color);
        }
        .footer_wrapper{
            display:flex;
            flex-direction:column-reverse;
            align-items:flex-start;
            padding: 20px;
            padding-bottom:0px !important;

        }

        .title{
            font-size: 40px;
            color:var(--primary_color);
            font-weight: 700;
            line-height: 36px;
            letter-spacing: -5px;
            margin-bottom: 10px;
        }
        .desc{
            font-size: 16px;
            color:var(--primary_color);
            line-height: 14px;
            letter-spacing: -2px;
            margin-bottom: 30px;
        }
        .footer_info_wrapper{
            margin-bottom:30px;
        }
        .robot_img_2 img{
            width:150px;
        }
        .pattern_img{
            display:none;
        }


        @container(min-width: 856px) {
            .footer_wrapper{
                gap:40px;
                flex-direction: row;
                align-items:flex-end;
                width:70%;
            }
            .pattern_img{
                display:block;
                position:absolute;
                top:0;
                right:0;
            }
            .pattern_img img{
                width: 300px;
                height: auto;
            }
        }

    `
}
