// timeline-filter/timeline-filter.js
const search_input = require('search-input')
const select_button = require('buttons/select-button')
const sm_icon_button = require('buttons/sm-icon-button')
const year_button = require('buttons/year-button')

const sheet = new CSSStyleSheet
sheet.replaceSync(get_theme())
/******************************************************************************
  TIMELINE FILTER COMPONENT
******************************************************************************/
// ----------------------------------------
// MODULE STATE & ID
var count = 0
const [cwd, dir] = [process.cwd(), __filename].map(x => new URL(x, 'file://').href)
const ID = dir.slice(cwd.length)
const STATE = { ids: {}, net: {} } // all state of component module
// ----------------------------------------
const default_opts = { }

module.exports = timeline_filter

function timeline_filter (opts = default_opts, protocol) {
  // ----------------------------------------
  // ID + JSON STATE
  // ----------------------------------------
  const name = 'timeline_filter-' + count++
  // ----------------------------------------
  // OPTS
  // ----------------------------------------
  const { img_src : { icon_arrow_up = `${prefix}/icon_arrow_up.svg` }} = opts.data
  // ----------------------------------------
  // PROTOCOL
  // ----------------------------------------
  const send = protocol({ from: name }, listen)
  const PROTOCOL = {}
  // ----------------------------------------
  // TEMPLATE
  // ----------------------------------------
  const el = document.createElement('div')
  const shadow = el.attachShadow( { mode:`closed` } )
  shadow.innerHTML = `<div class="filter_wrapper">
    <div class="timeline_filter">
      <div class="date_wrapper"></div>
    </div>
  </div>`
  shadow.adoptedStyleSheets = [sheet]
  const timeline_filter = shadow.querySelector('.timeline_filter')
  const date_wrapper = shadow.querySelector('.date_wrapper')
  // ----------------------------------------
  // ELEMENTS
  // ----------------------------------------
  const search_project = search_input(opts, timeline_filter_protocol)
  const status_button = select_button({ data: opts.data, name: 'STATUS', choices: ['ACTIVE', 'UNACTIVE', 'PAUSED'] }, timeline_filter_protocol)
  const tag_button = select_button({ data: opts.data, name: 'TAGS', choices: opts.tags }, timeline_filter_protocol)
  const month_button = sm_icon_button({ src: icon_arrow_up, activate: true })
  const year_btn = year_button({ data: opts.data, latest_date: opts.latest_date }, timeline_filter_protocol)
  timeline_filter.prepend(status_button, tag_button, search_project)
  date_wrapper.append(month_button, year_btn)
  // @TODO: change to month_protocol and year_protocol
  month_button.onclick = e => send({
      head: {by:name, to:'app_timeline', mid:0},
      type: 'toggle_month_filter',
      data: null
  })
  year_btn.onclick = e => send({
    head: { by: name, to: 'app_timeline', mid: 0 },
    type: 'toggle_year_filter',
    data: null
  })
  // ----------------------------------------
  // INIT
  // ----------------------------------------

  return el

  function timeline_filter_protocol (handshake, send, mid = 0) {
    if (handshake && handshake.from.includes('year_button')) PROTOCOL['get_date'] = send
    return listen
    function listen (message) {
      const { head,  refs, type, data, meta } = message
      const { by, to, id } = head
      message = { head: { by: name, to: 'app_timeline', mid: 0 }, type, data }
      send(message)
    }
  }
  function listen (message) {
    message.head.to = 'year_button'
    PROTOCOL['get_date'](message)
  }
}
function get_theme () {
  return `
    .filter_wrapper {
      container-type: inline-size;
      .timeline_filter {
        display: grid;
        grid-template-columns: 12fr;
        align-items: flex-end;   
        .date_wrapper {
          display: grid;
          grid-template-columns: 1fr 12fr;
        }
      }
    }
    @container (min-width: 450px) {
      .filter_wrapper {
        .timeline_filter {
          grid-template-columns: 1fr 1fr 9fr 1fr;
        }
      }
    }
  `
}
