// our-members/our-members.js
const window_bar = require('window-bar')

const sheet = new CSSStyleSheet
sheet.replaceSync(get_theme())
/******************************************************************************
  OUR MEMBERS COMPONENT
******************************************************************************/
// ----------------------------------------
// MODULE STATE & ID
var count = 0
const [cwd, dir] = [process.cwd(), __filename].map(x => new URL(x, 'file://').href)
const ID = dir.slice(cwd.length)
const STATE = { ids: {}, net: {} } // all state of component module
// ----------------------------------------
const default_opts = { }

module.exports = our_member

function our_member (opts = default_opts, protocol) {
  // ----------------------------------------
  // ID + JSON STATE
  // ----------------------------------------
  const name = `our_member`
  // ----------------------------------------
  // OPTS
  // ----------------------------------------
  const { data } = opts
  // Assigning all the icons
  const { img_src } = data
  const {
      icon_pdf_reader
  } = img_src
  // ----------------------------------------
  // PROTOCOL
  // ----------------------------------------
  protocol({ from: name }, listen)
  function listen () {
    our_member_wrapper.style.display = 'inline'
  }
  // ----------------------------------------
  // TEMPLATE
  // ----------------------------------------
  const el = document.createElement('div')
  const shopts = { mode: 'closed' }
  const shadow = el.attachShadow(shopts)
  shadow.adoptedStyleSheets = [sheet]
  shadow.innerHTML = `<div class="our_member">
    <div class="windowbar"></div>
    <div class="member_content">
      <h2>## our members</h2>
      <table>
        <thead>
          <tr>
            <td> s.no </td><td> names </td><td> socials </td>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td> 01 </td><td> alexander </td><td> cabal  </td>
          </tr>
          <tr>
            <td> 02 </td><td> alexander praetorius </td><td> geut/she </td>
          </tr>
        </tbody>
      </table>  
    </div>
  </div>`
  const our_member_wrapper = shadow.querySelector('.our_member')
  // ----------------------------------------
  const windowbar_shadow = shadow.querySelector('.windowbar').attachShadow(shopts)
  // ----------------------------------------
  // ELEMENTS
  // ----------------------------------------
  const window = window_bar({
    name:'our_member.md', 
    src: icon_pdf_reader,
    data: data
  }, our_member_protocol)
  windowbar_shadow.append(window)
  // ----------------------------------------
  // INIT
  // ----------------------------------------

  return el

  // cover protocol
  function our_member_protocol (message, send) {
    return listen
    // Listening to toggle event 
    function listen (message) {
      const { head, refs, type, data, meta } = message  
      const PROTOCOL = {
        'toggle_active_state': toggle_active_state
      }
      const action = PROTOCOL[type] || invalid      
      action(message)
    }
    function invalid (message) { console.error('invalid type', message) }
    async function toggle_active_state (message) {
      const { head, refs, type, data, meta } = message
      const { active_state } = data
      ;( active_state === 'active')?our_member_wrapper.style.display = 'none':''
    }
  }
}
function get_theme () {
  return `
    * {
      box-sizing: border-box;
    }
    .our_member {
      display: none;
      .member_content {
        position: relative;
        display: flex;
        flex-direction: column;
        width: 100vw;
        height: 100vh;
        padding: 10px;
        background-size: 10px 10px;
        background-color: var(--bg_color);
        border: 1px solid var(--primary_color);
        margin-bottom: 30px;
        h2 {
          margin: 0;
        }
        table {
          border-collapse: collapse;
          thead {
            font-weight: bold;
          }
          td {
            border: 1px solid var(--primary_color);
            padding: 8px;
          }
        }
      }
    }
    @container (min-width: 510px) {
      .our_member {
        .member_content {
          width: auto;
          height: auto;
        }
      }
    }
  `
}
