// timeline_card/index.js
module.exports = timeline_card


const path = require('path')
const cwd = process.cwd()
const prefix = path.relative(cwd, __dirname)


// CSS Boiler Plat
const sheet = new CSSStyleSheet
const theme = get_theme()
sheet.replaceSync(theme)





function timeline_card (opts) {

    const {data} = opts
    // Assigning all the icons
    const {img_src} = data
    const {
        icon_clock,
        icon_link,
        icon_calendar,
    } = img_src


    const el = document.createElement('div')
    el.style.lineHeight = '0px'
    const shadow = el.attachShadow( { mode : 'closed' } )
    
    shadow.innerHTML = `
        <div class="timeline_card">
            <div class="content_wrapper">

                <div class="icon_wrapper">
                    <div> ${icon_calendar} ${opts.date} </div>
                    <div> ${icon_clock} ${opts.time} </div>
                    <div> <a href="${opts.link}">${icon_link}</a> </div>
                </div>

                <div class="title"> ${opts.title} </div>
                <div class="desc"> ${opts.desc}</div>

            </div>
            <div class="tags_wrapper">
                ${opts.tags.map((tag) => `<div class="tag">${tag}</div>`).join('')}
            </div>
        </div>
        <style>${get_theme}</style>
    `


    shadow.adoptedStyleSheets = [sheet]
    return el


}





function get_theme(){
    return`
        *{
            box-sizing: border-box;
        }

        .timeline_card{
            height:max-content;
            width:100%;
            line-height: normal;
            background-color: var(--bg_color);
            color: var(--primary_color) !important;
            border:1px solid var(--primary_color);
            container-type: inline-size;
        }
        .content_wrapper{
            padding:20px;
        }
        .icon_wrapper{
            display:flex;
            gap:20px;
        }
        .icon_wrapper div{
            display:flex;
            gap:5px;
            font-size:16px;
            letter-spacing:-2px;
            align-items:center;
        }
        .icon_wrapper img{
            width: 20px;
            height: 20px;
        }
        .icon_wrapper div:nth-last-child(1){
            margin-left:auto;
        }
        .title{
            margin-top:20px;
            margin-bottom:5px;
            font-size:18px;
            font-weight: 700;
            letter-spacing: -2px;
            line-height:16px;
        }
        .desc{
            font-size:14px;
            letter-spacing:-2px;
            line-height:16px;
        }
        .tags_wrapper{
            display: flex;
            flex-wrap:wrap;
        }
        .tag{
            flex-grow:1;
            min-width:max-content;
            padding:5px 10px;
            border: 1px solid var(--primary_color);
            // line-height:0px;
            text-align:center;
        }
        




        @container(min-width: 856px) {
            
        }


    `
}
