// app_about_us/index.js
module.exports = app_about_us


const path = require('path')
const cwd = process.cwd()
const prefix = path.relative(cwd, __dirname)

const window_bar = require('window_bar')
const my_theme = require('my_theme')
const sm_text_button = require('buttons/sm_text_button')

// CSS Boiler Plat
const sheet = new CSSStyleSheet
const theme = get_theme()
sheet.replaceSync(theme)





function app_about_us () {

    // Assigning all the icons
    const { img_src: { 
        about_us_cover = `${prefix}/about_us_cover.png`,
        img_robot_1 = `${prefix}/img_robot_1.svg`,
        icon_pdf_reader = `${prefix}/icon_pdf_reader.svg`,
    } } = my_theme;

    const el = document.createElement('div')
    const shadow = el.attachShadow ( { mode : 'closed' } )

    shadow.innerHTML = `
        <div class="about_us_wrapper">
            <div class="about_us_cover_image"></div>
            <div class="content_wrapper">
                <div class="title"> ABOUT US </div>
            </div>
        </div>
        <div class="about_us_desc">
            Dat ecosystem garden supports open source projects that strengthen P2P foundations, with a focus on builder tools, infrastructure, research, and community resources.
        </div>
        <style> ${get_theme} </style>
    `

    // Listening to toggle event 
    const listen = (props) =>{
        const {active_state} = props
        ;(active_state==='active')?el.style.display = 'none':''
    }

    // Added background banner cover
    const about_us_cover_image = shadow.querySelector('.about_us_cover_image')
    banner_img = document.createElement('img')
    banner_img.src = about_us_cover
    about_us_cover_image.append(banner_img)

    // about_us_wrapper.style.backgroundImage = `url(${banner_cover})`
    const content_wrapper = shadow.querySelector('.content_wrapper')
    img_robot_1_img = document.createElement('img')
    img_robot_1_img.src = img_robot_1
    content_wrapper.prepend(img_robot_1_img)

    const view_more_btn = sm_text_button({text:'IMPORTANT DOCUMENTS'})
    const tell_me_more_btn = sm_text_button({text:'TELL ME MORE'})
    const cover_window = window_bar({
        name:'Learn_about_us.pdf', 
        src: icon_pdf_reader,
        action_buttons: [tell_me_more_btn, view_more_btn]
    }, listen)

    shadow.adoptedStyleSheets = [ sheet ]
    shadow.prepend(cover_window)
    return el

}




function get_theme(){
    return`
        *{
            box-sizing: border-box;
        }

        .about_us_wrapper{
            position:relative;
            height:max-content;
            width:100%;
            display:flex;
            flex-direction:column;
            justify-content: center;
            align-items: center;
            padding: 150px 0px;
            background-image: radial-gradient(#A7A6A4 1px, #FFF 1px);
            background-size: 10px 10px;
            background-color:red;
            border: 1px solid var(--primary_color);
            box-sizing: border-box;
            container-type: inline-size;
        }

        /* This covers background-image will change to an image */
        .about_us_cover_image{
            position: absolute;
            width:100%;
            height:100%;
            overflow:hidden;
        }
        .about_us_cover_image img{
            position:absolute;
            left:50%;
            top:50%;
            width: auto;
            height: 80%;
            transform:translate(-50%, -50%);
        }

        .about_us_desc{
            width:100% !important;
            background-color:var(--bg_color);
            color: var(--primary_color);
            border:1px solid var(--primary_color);
            padding:10px;
            letter-spacing: -2px;
            line-height:18px;
            font-size:16px;
            margin-bottom:30px;
            box-sizing: border-box;
        }


        /* Cover image alignment */
        .content_wrapper{
            display: flex;
            justify-content:center;
            align-items:center;
            gap:20px;
            position: relative;
            z-index:1;
            color:var(--primary_color);
            text-align:center;
        }
        .content_wrapper img{
            width: 100px;
            height: auto;
        }
        .content_wrapper .title{
            font-size:40px;
        }


        @container(min-width: 856px) {
            .about_us_cover_image img{
                width: 100%;
                height: auto;
            }
        }

    `
}
