// consortium_page/index.js
module.exports = consortium_page

const mission_statement = require('mission_statement')

function consortium_page (opts, protocol) {

    //Image data
    const {data} = opts
    const {img_src} = data
    const {icon_pdf_reader} = img_src

    //Communication
    const PROTOCOL = {}

    const el = document.createElement('div')
    const shadow = el.attachShadow({mode: 'closed'})

    // adding a `main_wrapper` 
    shadow.innerHTML = `
        <div class="main_wrapper">
            <div class="icon_wrapper">
            </div>
            <div class="popup_wrapper">
            </div>
        </div>
        <style>${get_theme()}</style>
    `
    const icon_wrapper = shadow.querySelector('.icon_wrapper')
    icon_wrapper.innerHTML = `
        ${icon_pdf_reader}
        <span>MISSION STATEMENT.MD</span>
    `
    icon_wrapper.ondblclick = () => {PROTOCOL.notify()}


    const popup_wrapper = shadow.querySelector('.popup_wrapper')
    popup_wrapper.append(mission_statement({data}, consortium_protocol))

    const main_wrapper = shadow.querySelector('.main_wrapper')
    main_wrapper.append(icon_wrapper, popup_wrapper)
    
    // shadow.adoptedStyleSheets = [sheet]
    return el

    function consortium_protocol (handshake, send, mid=0){
        PROTOCOL['notify'] = send
    }
}

function get_theme() {
    return `
        .main_wrapper{
            display: flex;
            justify-content: space-between;
            margin: 0;
            padding:30px 10px;
            opacity: 1;
            background-image: radial-gradient(var(--primary_color) 2px, var(--bg_color) 2px);
            background-size: 16px 16px;
        }
        .icon_wrapper{
            display: flex;
            flex-direction: column;
            width: fit-content;
            height: fit-content;
            align-items: center;
            user-select: none;
        }
        .icon_wrapper:hover{
            cursor: default;
        }
        .icon_wrapper span{
            background-color: white;
            width: 150px;
            text-align: center;
        }
        svg{
            height: 50px;
            width: 50px;
            margin: 5px 0;
            background-color: white;

        }
        svg path{
            fill: black;
        }
    `
}
