// projects-page/projects-page.js
const app_projects = require('app-projects')
const the_dat = require('the-dat')
const app_footer = require('app-footer')

const sheet = new CSSStyleSheet
sheet.replaceSync(get_theme())
/******************************************************************************
  PROJECTS PAGE COMPONENT
******************************************************************************/
// ----------------------------------------
// MODULE STATE & ID
var count = 0
const [cwd, dir] = [process.cwd(), __filename].map(x => new URL(x, 'file://').href)
const ID = dir.slice(cwd.length)
const STATE = { ids: {}, net: {} } // all state of component module
// ----------------------------------------
const default_opts = { }

module.exports = projects_page

function projects_page (opts = default_opts, protocol) {
  // ----------------------------------------
  // ID + JSON STATE
  // ----------------------------------------
  const id = count++
  // ----------------------------------------
  // OPTS
  // ----------------------------------------
  const { data } = opts
  // ----------------------------------------
  // PROTOCOL
  // ----------------------------------------
  const PROTOCOL = {}
  // ----------------------------------------
  // TEMPLATE
  // ----------------------------------------
  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })
  // adding a `main-wrapper` 
  shadow.innerHTML = `<div class="main-wrapper">
    <div class="main"></div>
  </div>`
  shadow.adoptedStyleSheets = [sheet]
  const main = shadow.querySelector('.main')
  // ----------------------------------------
  // ELEMENTS
  // ----------------------------------------
  const components = [
    the_dat({ data }, projects_protocol),
    app_projects({ data }),
    app_footer({ data })
  ]
  main.append(...components)
  // ----------------------------------------
  // INIT
  // ----------------------------------------

  return el

  // Placeholder code for learning purposes
  // Will be removed
  function projects_protocol (handshake, send){
    if (send) return listen
    const PROTOCOL = {
      'toggle_display' : toggle_display
    }
    send = handshake(null, listen)
    function listen (message){
      function format (new_message = {
        head: [from = 'alice', to = 'bob', message_id = 1],
        refs: { cause: message.head }, // reply to received message
        type: 'change_theme',
        data: `.foo { background-color: red; }`
      }) { return new_message }
      console.log(format())
      // const { head, type, data } = message
      // const [by, to, id] = head
      // if (to !== id) return console.error('address unknown', message)
      // const action = PROTOCOL[type] || invalid
      // action(message)
    }
    function invalid (message) { console.error('invalid type', message) }
    async function toggle_display ({ head: [to], data: theme }) {
      // @TODO: apply theme to `sheet` and/or `style` and/or css `var(--property)`
    }
  }
}
function get_theme () {
  return `
    * {
      box-sizing: border-box;
    }
    .main-wrapper {
      container-type: inline-size;
      .main {
        margin: 0;
        padding: 30px 10px;
        opacity: 1;
        background-image: radial-gradient(var(--primary_color) 2px, var(--bg_color) 2px);
        background-size: 16px 16px;
      }
    }
    @container (min-width: 856px) {
      .main {
        padding-inline: 20px !important;
      }
    }
  `
}
