// important-documents/important-documents.js
const window_bar = require('window-bar')
/******************************************************************************
  IMPORTANT DOCUMENTS COMPONENT
******************************************************************************/
// ----------------------------------------
// MODULE STATE & ID
var count = 0
const [cwd, dir] = [process.cwd(), __filename].map(x => new URL(x, 'file://').href)
const ID = dir.slice(cwd.length)
const STATE = { ids: {}, net: {} } // all state of component module
// ----------------------------------------
const sheet = new CSSStyleSheet
sheet.replaceSync(get_theme())
const default_opts = { }
const shopts = { mode: 'closed' }
// ----------------------------------------
module.exports = important_documents
// ----------------------------------------
function important_documents (opts = default_opts, protocol) {
  // ----------------------------------------
  // ID + JSON STATE
  // ----------------------------------------
  const id = `${ID}:${count++}` // assigns their own name
  const status = {}
  const state = STATE.ids[id] = { id, status, wait: {}, net: {}, aka: {} } // all state of component instance
  const cache = resources({})
  // ----------------------------------------
  // OPTS
  // ----------------------------------------
  const { data } = opts
  // Assigning all the icons
  const { img_src } = data
  const {
    icon_pdf_reader
  } = img_src
  // ----------------------------------------
  // PROTOCOL
  // ----------------------------------------
  const on = { 'show': on_show, 'hide': on_hide }
  const channel = use_protocol('up')({ protocol, state, on })
  // ----------------------------------------
  // TEMPLATE
  // ----------------------------------------
  const el = document.createElement('div')
  const shadow = el.attachShadow(shopts)
  shadow.adoptedStyleSheets = [sheet]
  shadow.innerHTML = `<div class="important_documents">
    <div class="window_bar_wrapper"></div>
    <div class="documents_content">
      <h2>Visit links for more info</h2>
      <ol type="1">
        <li> <a href="https://github.com/dat-ecosystem/organization/blob/main/code-of-conduct.md" target="_blank">code of conduct</a> </li>
        <li> <a href="https://github.com/dat-ecosystem/organization/blob/main/code-of-conduct-contact.md" target="_blank">code of conduct - reporting guide</a> </li>
        <li> <a href="https://github.com/dat-ecosystem/comm-comm#readme" target="_blank">comm comm calls</a> </li>
        <li> <a href="https://github.com/dat-ecosystem/organization/tree/main/consortium/meeting%20notes" target="_blank">consortium meeting notes</a> </li>
        <li> <a href="https://github.com/dat-ecosystem/organization/tree/main/assets/logo%20%26%20visuals%20final" target="_blank">visuals</a> </li>
        <li> <a href="https://github.com/dat-ecosystem/organization" target="_blank">organization repository</a> </li>
      </ol>
    </div>
  </div>`
  const important_documents_wrapper = shadow.querySelector('.important_documents')
  // ----------------------------------------
  const window_bar_wrapper = shadow.querySelector('.window_bar_wrapper').attachShadow(shopts)
  // ----------------------------------------
  // ELEMENTS
  // ----------------------------------------
  { // windowbar
    const on = {
      'toggle_active_state': toggle_active_state
    }
    const protocol = use_protocol('windowbar')({ state, on })
    const opts = {
      name: 'important_documents.md', 
      src: icon_pdf_reader,
      data
    }
    const element = window_bar(opts, protocol)
    window_bar_wrapper.append(element)
    async function toggle_active_state (message) {
      const { active_state } = message.data
      if (active_state === 'active') important_documents_wrapper.style.display = 'none'
      const channel = state.net[state.aka.up]
      channel.send({
        head: [id, channel.send.id, channel.mid++],
        type: 'deactivate_tick',
        data: opts.name
      })
    }
  }
  // ----------------------------------------
  // INIT
  // ----------------------------------------

  return el

  function on_show (message) {
    important_documents_wrapper.style.display = 'inline'
  }
  function on_hide (message) {
    important_documents_wrapper.style.display = 'none'
  }
}
function get_theme () {
  return `
    * {
      box-sizing: border-box;
    }
    .important_documents {
      display: inline;
    }
    .important_documents .documents_content {
      position: relative;
      display: flex;
      width: 92vw;
      height: 100vh;
      flex-direction: column;
      padding: 10px;
      background-size: 10px 10px;
      background-color: var(--bg_color);
      border: 1px solid var(--primary_color);
      margin-bottom: 30px;
    }
    .important_documents .documents_content h2 {
      margin: 0;
    }
    .important_documents a{
      color: var(--primary_color);
      text-decoration: none;
    }
    .important_documents a:hover{
      text-decoration: underline;
    }
    @container (min-width: 510px) {
      .important_documents .documents_content {
        width: auto;
        height: auto;
      }
    }
  `
}
// ----------------------------------------------------------------------------
function shadowfy (props = {}, sheets = []) {
  return element => {
    const el = Object.assign(document.createElement('div'), { ...props })
    const sh = el.attachShadow(shopts)
    sh.adoptedStyleSheets = sheets
    sh.append(element)
    return el
  }
}
function use_protocol (petname) {
  return ({ protocol, state, on = { } }) => {
    if (petname in state.aka) throw new Error('petname already initialized')
    const { id } = state
    const invalid = on[''] || (message => console.error('invalid type', message))
    if (protocol) return handshake(protocol(Object.assign(listen, { id })))
    else return handshake
    // ----------------------------------------
    // @TODO: how to disconnect channel
    // ----------------------------------------
    function handshake (send) {
      state.aka[petname] = send.id
      const channel = state.net[send.id] = { petname, mid: 0, send, on }
      return protocol ? channel : Object.assign(listen, { id })
    }
    function listen (message) {
      const [from] = message.head
      const by = state.aka[petname]
      if (from !== by) return invalid(message) // @TODO: maybe forward
      console.log(`[${id}]:${petname}>`, message)
      const { on } = state.net[by]
      const action = on[message.type] || invalid
      action(message)
    }
  }
}
// ----------------------------------------------------------------------------
function resources (pool) {
  var num = 0
  return factory => {
    const prefix = num++
    const get = name => {
      const id = prefix + name
      if (pool[id]) return pool[id]
      const type = factory[name]
      return pool[id] = type()
    }
    return Object.assign(get, factory)
  }
}
