// app_timeline_mini/index.js
module.exports = app_timeline_mini


const path = require('path')
const cwd = process.cwd()
const prefix = path.relative(cwd, __dirname)

const window_bar = require('window_bar')
const timeline_card = require('timeline_card')
const sm_text_button = require('buttons/sm_text_button')
const scrollbar = require('scrollbar')

// CSS Boiler Plat
const sheet = new CSSStyleSheet
const theme = get_theme()
sheet.replaceSync(theme)





function app_timeline_mini (opts) {

    console.log(opts)
    const {data} = opts

    // Assigning all the icons
    const { img_src: {
        icon_folder= `${prefix}/icon_folder.svg`,
    } } = data

    const el = document.createElement('div')
    const shadow = el.attachShadow ( { mode : 'closed' } )

    shadow.innerHTML = `
        <div class="main_wrapper">
            <div class="timeline_wrapper">
            </div>
        </div>
        <style> ${get_theme} </style>
    `

    // Listening to toggle event 
    const listen = (props) =>{
        const {active_state} = props
        ;(active_state==='active')?el.style.display = 'none':''
    }

    // Adding Applicatin window Bar
    const view_more_btn = sm_text_button({text:'View more (12)'})
    const cover_window = window_bar({
        name:'TIMELINE', 
        src: icon_folder,
        action_buttons: [view_more_btn],
        data: data
    }, listen)


    // Adding timeline cards
    const timeline_wrapper = shadow.querySelector('.timeline_wrapper')
    const cards_data = [
        { title: 'Official starting of the web course.', data: data,  date: 'July 11, 2022', time: '07:05AM', link: '/', desc: 'The course is called - vanilla.js hyper modular web component building course and it will last approximately 4-8 weeks.. ', tags: ['Hypercore', 'Hypercore', 'Hypercore'],
        },{ title: 'Official starting of the web course.', data: data,  date: 'July 11, 2022', time: '07:05AM', link: '/', desc: 'The course is called - vanilla.js hyper modular web component building course and it will last approximately 4-8 weeks.. ', tags: ['Hypercore', 'Hypercore', 'Hypercore'],
        },{ title: 'Official starting of the web course.', data: data,  date: 'July 11, 2022', time: '07:05AM', link: '/', desc: 'The course is called - vanilla.js hyper modular web component building course and it will last approximately 4-8 weeks.. ', tags: ['Hypercore', 'Hypercore', 'Hypercore'],
        },{ title: 'Official starting of the web course.', data: data,  date: 'July 11, 2022', time: '07:05AM', link: '/', desc: 'The course is called - vanilla.js hyper modular web component building course and it will last approximately 4-8 weeks.. ', tags: ['Hypercore', 'Hypercore', 'Hypercore'],
        },{ title: 'Official starting of the web course.', data: data,  date: 'July 11, 2022', time: '07:05AM', link: '/', desc: 'The course is called - vanilla.js hyper modular web component building course and it will last approximately 4-8 weeks.. ', tags: ['Hypercore', 'Hypercore', 'Hypercore'],
        },{ title: 'Official starting of the web course.', data: data,  date: 'July 11, 2022', time: '07:05AM', link: '/', desc: 'The course is called - vanilla.js hyper modular web component building course and it will last approximately 4-8 weeks.. ', tags: ['Hypercore', 'Hypercore', 'Hypercore'],
        },{ title: 'Official starting of the web course.', data: data,  date: 'July 11, 2022', time: '07:05AM', link: '/', desc: 'The course is called - vanilla.js hyper modular web component building course and it will last approximately 4-8 weeks.. ', tags: ['Hypercore', 'Hypercore', 'Hypercore'],
        },{ title: 'Official starting of the web course.', data: data,  date: 'July 11, 2022', time: '07:05AM', link: '/', desc: 'The course is called - vanilla.js hyper modular web component building course and it will last approximately 4-8 weeks.. ', tags: ['Hypercore', 'Hypercore', 'Hypercore'],
        },{ title: 'Official starting of the web course.', data: data,  date: 'July 11, 2022', time: '07:05AM', link: '/', desc: 'The course is called - vanilla.js hyper modular web component building course and it will last approximately 4-8 weeks.. ', tags: ['Hypercore', 'Hypercore', 'Hypercore'],
        },{ title: 'Official starting of the web course.', data: data,  date: 'July 11, 2022', time: '07:05AM', link: '/', desc: 'The course is called - vanilla.js hyper modular web component building course and it will last approximately 4-8 weeks.. ', tags: ['Hypercore', 'Hypercore', 'Hypercore'],
        },
    ]
    const timeline_cards = cards_data.map((card_data) => timeline_card(card_data))
    timeline_cards.forEach((card) => {
        timeline_wrapper.append(card)
    })

    const main_wrapper = shadow.querySelector('.main_wrapper')
    
    
    let state;
    function protocol(message, notify){
        timeline_wrapper.onscroll = message;
        state = notify
        return [listen1, setScrollTop]
    }
    function listen1 (){
        state({sh: timeline_wrapper.scrollHeight, ch: timeline_wrapper.clientHeight, st: timeline_wrapper.scrollTop})
    }
    function setScrollTop(value){
        timeline_wrapper.scrollTop = value
    }

    const sc = scrollbar({data: data}, protocol);
    main_wrapper.append(sc)
    
    shadow.adoptedStyleSheets = [ sheet ]
    shadow.prepend(cover_window)
    return el

}




function get_theme(){
    return`
        *{
            box-sizing: border-box;
        }
        .timeline_wrapper{
            --s: 20px; /* control the size */
            --_g: var(--bg_color) /* first color */ 0 25%, #0000 0 50%;
            background:
                repeating-conic-gradient(at 66% 66%,var(--_g)),
                repeating-conic-gradient(at 33% 33%,var(--_g)),
                var(--primary_color);  /* second color */ 
            background-size: var(--s) var(--s);  
            overflow: scroll;
            border:1px solid var(--primary_color);
            width:100%;
            height: 400px;
            padding: 0px;
            display: grid;
            gap:20px;
            grid-template-columns: 12fr;
        }
        .timeline_wrapper::-webkit-scrollbar {
            -webkit-appearance: none;
        }
        @container(min-width: 768px) {
            .timeline_wrapper{
                grid-template-columns: repeat(2, 6fr);
            }
        }
    
        @container(min-width: 1200px) {
            .timeline_wrapper{
                grid-template-columns: repeat(3, 4fr);
            }
        }
        .main_wrapper{
            display: flex;
            container-type: inline-size;
            width: 100%;
            height: 100%;
            margin-bottom: 30px;
        }
          
        /*---------- Mobile devices ----------*/
        @media (min-width: 480px) {
        }

        /*---------- iPads, Tablets ----------*/
        @media (min-width: 768px) {
        }

        /*---------- Mediuem screens, laptops ----------*/
        @media (min-width: 1024px) {}
    `
}
