// month_card/index.js
module.exports = month_card

const sheet = new CSSStyleSheet
const theme = get_theme()
sheet.replaceSync(theme)

const day_button = require('buttons/day_button')

let id = 0;

function month_card (opts, protocol) {
    const name = `month_card-${id++}`
    const notify = protocol({from: name}, listen)

    const el = document.createElement('div')
    const shadow = el.attachShadow({mode: 'closed'})

    shadow.innerHTML = `
        <div class="month_card">
            <span class="month_name"><b>${opts.name}</b></span>
            <div class="days_wrapper"></div>
        </div>
        <style>${get_theme}</style>
    `
    const days_wrapper = shadow.querySelector('.days_wrapper')
    for(let i=1; i<=opts.days; i++){
        const btn = day_button()
        btn.id = i
        btn.onclick = e => {
            notify({
                head: {by:name, to:'month_filter', mid: 0},
                type: 'toggle_day_button',
                data: e.target.id
            })
        }
        days_wrapper.append(btn)
    }
    
    const month_name = shadow.querySelector('.month_name')
    month_name.onclick = e => {
        notify({
            head: {by:name, to:'month_filter', mid: 0},
            type: 'toggle_active_state',
            data: opts.name
        })
    }

    shadow.adoptedStyleSheets = [sheet]
    return el

    function listen (message){
        month_name.classList.toggle('active')

    }
}

function get_theme() {
    return `
    .month_card {
        width: 140px;
        height: 130px;
        border: 1px solid var(--primary_color);
        border-right-width: 4px;
        background-color: var(--bg_color);
        margin-top: -1px;
        margin-left: -1px;
    }
    .month_name {
        display: block;
        text-align: center;
        padding: 5px 0;
        cursor: pointer;
    }
    .month_name.active{
        background-color: var(--ac-1)
    }
    .days_wrapper {
        display: flex;
        flex-wrap: wrap;
        border-top: 1px solid var(--primary_color);
        
    }
    `
}
