// our-alumni/our-alumni.js
const window_bar = require('window-bar')
const scrollbar = require('scrollbar')
const alumni = require('./alumni.json');

/******************************************************************************
  OUR alumni COMPONENT
******************************************************************************/
// ----------------------------------------
// MODULE STATE & ID
var count = 0
const [cwd, dir] = [process.cwd(), __filename].map(x => new URL(x, 'file://').href)
const ID = dir.slice(cwd.length)
const STATE = { ids: {}, net: {} } // all state of component module
// ----------------------------------------
const sheet = new CSSStyleSheet
sheet.replaceSync(get_theme())
const default_opts = { }
const shopts = { mode: 'closed' }
// ----------------------------------------
module.exports = our_alumni
// ----------------------------------------
function our_alumni (opts = default_opts, protocol) {
  // ----------------------------------------
  // RESOURCE POOL (can't be serialized)
  // ----------------------------------------
  const ro = new ResizeObserver(entries => {
    console.log('ResizeObserver:terminal:resize')
    const scroll_channel = state.net[state.aka.scrollbar]
    scroll_channel.send({
      head: [id, scroll_channel.send.id, scroll_channel.mid++],
      refs: { },
      type: 'handle_scroll',
    })
  })
  // ----------------------------------------
  // ID + JSON STATE
  // ----------------------------------------
  const id = `${ID}:${count++}` // assigns their own name
  const status = {}
  const state = STATE.ids[id] = { id, status, wait: {}, net: {}, aka: {} } // all state of component instance
  const cache = resources({})
  // ----------------------------------------
  // OPTS
  // ----------------------------------------
  const { data } = opts
  // Assigning all the icons
  const { img_src } = data
  const {
    icon_pdf_reader
  } = img_src
  // ----------------------------------------
  // PROTOCOL
  // ----------------------------------------
  const on = { 'show': on_show, 'hide': on_hide }
  const channel = use_protocol('up')({ protocol, state, on })
  function on_show (message) {
    our_alumni_wrapper.style.display = 'block'
  }
  function on_hide (message) {
    our_alumni_wrapper.style.display = 'none'
  }
  // ----------------------------------------
  // TEMPLATE
  // ----------------------------------------
  const el = document.createElement('div')
  const shadow = el.attachShadow(shopts)
  shadow.adoptedStyleSheets = [sheet]
  shadow.innerHTML = `<div class="our_alumni">
    <div class="windowbar"></div>
    <div class="scrollbar_wrapper">
      <div class="alumni_content">
        <h2>## our alumni</h2>
      </div>
    </div>
  </div>`

  // Function to create a table from the alumni data
  function createTable(data) {
    const tableHeader = '<thead><tr><td> no. </td><td> name </td></tr></thead>'
    const tableRows = data.map((alumni, index) => {
      const anchorLink = `<a href="${alumni.link}" target="_blank">${alumni.name}</a>`
      return `<tr><td> ${index + 1} </td><td> ${anchorLink} </td></tr>`
    })
    const table = `<table>${tableHeader}<tbody>${tableRows}</tbody></table>`
    return table
  }

  const tableHTML = createTable(alumni)
  const tempContainer = document.createElement('div')
  tempContainer.innerHTML = tableHTML
  const tableElement = tempContainer.querySelector('table')
  const alumni_content = shadow.querySelector('.alumni_content')
  alumni_content.appendChild(tableElement)



  const our_alumni_wrapper = shadow.querySelector('.our_alumni')
  const scrollbar_wrapper = shadow.querySelector('.scrollbar_wrapper')
  // ----------------------------------------
  const windowbar_shadow = shadow.querySelector('.windowbar').attachShadow(shopts)
  // ----------------------------------------
  // ELEMENTS
  // ----------------------------------------
  { // windowbar
    const on = {
      'toggle_active_state': toggle_active_state
    }
    const protocol = use_protocol('windowbar')({ state,  on })
    const opts = {
      name:'our_alumni.md', 
      src: icon_pdf_reader,
      data
    }
    const element = window_bar(opts, protocol)
    windowbar_shadow.append(element)
    async function toggle_active_state (message) {
      const { active_state } = message.data
      if (active_state === 'active') our_alumni_wrapper.style.display = 'none'
      const channel = state.net[state.aka.up]
      channel.send({
        head: [id, channel.send.id, channel.mid++],
        type: 'deactivate_tick',
        data: opts.name
      })
    }
  }
  { // scrollbar
    const on = { 'set_scroll': on_set_scroll, status: onstatus }
    const protocol = use_protocol('scrollbar')({ state, on })
    opts.data.img_src.icon_arrow_start = opts.data.img_src.icon_arrow_up
    opts.data.img_src.icon_arrow_end = opts.data.img_src.icon_arrow_down
    const opts1 = { data }
    const element = shadowfy()(scrollbar(opts1, protocol))
    scrollbar_wrapper.append(element)
    alumni_content.onscroll = on_scroll
  }
  // ----------------------------------------
  // INIT
  // ----------------------------------------
  watch_scrollbar()

  return el

  function watch_scrollbar () {
    const channel = state.net[state.aka.scrollbar]
    ro.observe(alumni_content)
  }
  function on_scroll (message) {
    const channel = state.net[state.aka.scrollbar]
    channel.send({
      head: [id, channel.send.id, channel.mid++],
      refs: { },
      type: 'handle_scroll',
    })
  }
  function on_set_scroll (message) {
    console.log('set_scroll', message) 
    setScrollTop(message.data)
  }
  function onstatus (message) {
    const channel = state.net[state.aka.scrollbar]
    channel.send({
      head: [id, channel.send.id, channel.mid++],
      refs: { cause: message.head },
      type: 'update_size',
      data: {
        sh: alumni_content.scrollHeight,
        ch: alumni_content.clientHeight,
        st: alumni_content.scrollTop
      }
    })
  }
  async function setScrollTop (value) {
    alumni_content.scrollTop = value
  }
}
function get_theme () {
  return `
    * {
      box-sizing: border-box;
    }
    .our_alumni {
      display: none;
      margin-bottom: 30px;
    }
    .alumni_content {
      position: relative;
      display: flex;
      flex-direction: column;
      width: 100%;
      height: 100%;
      padding: 10px;
      background-size: 10px 10px;
      background-color: var(--bg_color);
      border: 1px solid var(--primary_color);
      max-height: 600px;
      overflow-y: scroll;
    }
    .alumni_content::-webkit-scrollbar{
      display: none;
    }
    .alumni_content h2 {
      margin: 0;
    }
    .alumni_content table {
      border-collapse: collapse;
    }
    .alumni_content table thead {
      font-weight: bold;
    }
    .alumni_content table td {
      border: 1px solid var(--primary_color);
      padding: 8px;
    }
    .our_alumni .scrollbar_wrapper{
      display: flex;
    }
  `
}
// ----------------------------------------------------------------------------
function shadowfy (props = {}, sheets = []) {
  return element => {
    const el = Object.assign(document.createElement('div'), { ...props })
    const sh = el.attachShadow(shopts)
    sh.adoptedStyleSheets = sheets
    sh.append(element)
    return el
  }
}
function use_protocol (petname) {
  return ({ protocol, state, on = { } }) => {
    if (petname in state.aka) throw new Error('petname already initialized')
    const { id } = state
    const invalid = on[''] || (message => console.error('invalid type', message))
    if (protocol) return handshake(protocol(Object.assign(listen, { id })))
    else return handshake
    // ----------------------------------------
    // @TODO: how to disconnect channel
    // ----------------------------------------
    function handshake (send) {
      state.aka[petname] = send.id
      const channel = state.net[send.id] = { petname, mid: 0, send, on }
      return protocol ? channel : Object.assign(listen, { id })
    }
    function listen (message) {
      const [from] = message.head
      const by = state.aka[petname]
      if (from !== by) return invalid(message) // @TODO: maybe forward
      console.log(`[${id}]:${petname}>`, message)
      const { on } = state.net[by]
      const action = on[message.type] || invalid
      action(message)
    }
  }
}
// ----------------------------------------------------------------------------
function resources (pool) {
  var num = 0
  return factory => {
    const prefix = num++
    const get = name => {
      const id = prefix + name
      if (pool[id]) return pool[id]
      const type = factory[name]
      return pool[id] = type()
    }
    return Object.assign(get, factory)
  }
}
