// project-filter/project-filter.js
const search_input = require('search-input')
const select_button = require('buttons/select-button')
/******************************************************************************
  WINDOW BAR COMPONENT
******************************************************************************/
// ----------------------------------------
// MODULE STATE & ID
var count = 0
const [cwd, dir] = [process.cwd(), __filename].map(x => new URL(x, 'file://').href)
const ID = dir.slice(cwd.length)
const STATE = { ids: {}, net: {} } // all state of component module
// ----------------------------------------
const sheet = new CSSStyleSheet
sheet.replaceSync(get_theme())
const default_opts = { }
const shopts = { mode: 'closed' }
// ----------------------------------------
module.exports = project_filter
// ----------------------------------------
function project_filter (opts = default_opts, protocol) {
  // ----------------------------------------
  // ID + JSON STATE
  // ----------------------------------------
  const id = `${ID}:${count++}` // assigns their own name
  const status = {}
  const state = STATE.ids[id] = { id, status, wait: {}, net: {}, aka: {} } // all state of component instance
  const cache = resources({})
  // ----------------------------------------
  // OPTS
  // ----------------------------------------
  const { tags, data } = opts
  // ----------------------------------------
  // PROTOCOL
  // ----------------------------------------
  const on = {}
  const up_channel = use_protocol('up')({ protocol, state, on })
  // ----------------------------------------
  // TEMPLATE
  // ----------------------------------------
  const el = document.createElement('div')
  const shadow = el.attachShadow(shopts)
  shadow.adoptedStyleSheets = [sheet]
  shadow.innerHTML = `<div class="filter_wrapper">
    <div class="project_filter"></div>
  </div>`
  const project_filter = shadow.querySelector('.project_filter')
  // ----------------------------------------
  // ELEMENTS
  // ----------------------------------------
  { // status button
    const on = { 'value': on_value }
    const protocol = use_protocol('status_button')({ state, on })
    const opts = { data, name: 'STATUS', choices: ['ACTIVE', 'INACTIVE', 'PAUSED'] }
    const element = shadowfy()(select_button(opts, protocol))
    project_filter.append(element)
  }
  { // tag button
    const on = { 'value': on_value }
    const protocol = use_protocol('tag_button')({ state, on })
    const opts = { data, name: 'TAGS', choices: tags }
    const element = shadowfy()(select_button(opts, protocol))
    project_filter.append(element)
  }
  { // search project
    const on = { 'value': on_value }
    const protocol = use_protocol('search_project')({ state, on })
    const search_opts = opts
    const element = shadowfy()(search_input(search_opts, protocol))
    project_filter.append(element)
  }
  // ----------------------------------------
  // INIT
  // ----------------------------------------

  return el

  function on_value (message) {
    up_channel.send({
      head: [id, up_channel.send.id, up_channel.mid++],
      refs: { cause: message.head },
      type: 'value',
      data: message.data
    })
  }
}
function get_theme () {
  return `
    .filter_wrapper {
      container-type: inline-size;
    }
    .project_filter {
      display: grid;
      grid-template-columns: 12fr;
      align-items: flex-end;
    }
    @container (min-width: 412px) {
      .project_filter {
        grid-template-columns: 1fr 1fr 10fr;
      }
    }
  `
} 
// ----------------------------------------------------------------------------
function shadowfy (props = {}, sheets = []) {
  return element => {
    const el = Object.assign(document.createElement('div'), { ...props })
    const sh = el.attachShadow(shopts)
    sh.adoptedStyleSheets = sheets
    sh.append(element)
    return el
  }
}
function use_protocol (petname) {
  return ({ protocol, state, on = { } }) => {
    if (petname in state.aka) throw new Error('petname already initialized')
    const { id } = state
    const invalid = on[''] || (message => console.error('invalid type', message))
    if (protocol) return handshake(protocol(Object.assign(listen, { id })))
    else return handshake
    // ----------------------------------------
    // @TODO: how to disconnect channel
    // ----------------------------------------
    function handshake (send) {
      state.aka[petname] = send.id
      const channel = state.net[send.id] = { petname, mid: 0, send, on }
      return protocol ? channel : Object.assign(listen, { id })
    }
    function listen (message) {
      const [from] = message.head
      const by = state.aka[petname]
      if (from !== by) return invalid(message) // @TODO: maybe forward
      console.log(`[${id}]:${petname}>`, message)
      const { on } = state.net[by]
      const action = on[message.type] || invalid
      action(message)
    }
  }
}
// ----------------------------------------------------------------------------
function resources (pool) {
  var num = 0
  return factory => {
    const prefix = num++
    const get = name => {
      const id = prefix + name
      if (pool[id]) return pool[id]
      const type = factory[name]
      return pool[id] = type()
    }
    return Object.assign(get, factory)
  }
}
