// my_theme/dark_theme/index.js
const path = require('path')
const cwd = process.cwd()
const prefix = path.relative(cwd, __dirname)


const my_theme = {
  img_src:{
    icon_consortium: `${prefix}/icon_consortium_page.png`,
    icon_blogger: `${prefix}/icon_blogger.png`,
    icon_discord: `${prefix}/icon_discord.png`,
    icon_twitter: `${prefix}/icon_twitter.png`,
    icon_github: `${prefix}/icon_github.png`,
    icon_terminal: `${prefix}/icon_terminal.png`,
    icon_theme: `${prefix}/icon_theme.png`,
    icon_close_dark: `${prefix}/icon_close_dark.svg`,
    icon_close_light: `${prefix}/icon_close_light.svg`,
    icon_pdf_reader: `${prefix}/icon_pdf_reader.svg`,
    icon_folder: `${prefix}/icon_folder.svg`,
    icon_arrow_down: `${prefix}/icon_arrow_down.svg`,
    icon_arrow_up: `${prefix}/icon_arrow_up.svg`,
    icon_arrow_down_light: `${prefix}/icon_arrow_down_light.svg`,
    icon_arrow_up_light: `${prefix}/icon_arrow_up_light.svg`,
    icon_arrow_up_light: `${prefix}/icon_arrow_up_light.svg`,
    icon_search : `${prefix}/icon_search.svg`,
    banner_cover : `${prefix}/banner_cover.svg`,
    about_us_cover : `${prefix}/about_us_cover.png`,
    tree_character : `${prefix}/tree_character.png`,
    icon_clock : `${prefix}/icon_clock.svg`,
    icon_link : `${prefix}/icon_link.svg`,
    icon_calendar : `${prefix}/icon_calendar.svg`,
    project_logo_1 : `${prefix}/project_logo_1.png`,
    img_robot_1 : `${prefix}/img_robot_1.png`,
    img_robot_2 : `${prefix}/img_robot_2.png`,
    pattern_img_1 : `${prefix}/pattern_img_1.png`,
  },
  light_theme:{
    bg_color : '#fff',
    primary_color : '#293648',
    ac_1 : '#2ACA4B',
    ac_2 : '#F9A5E4',
    ac_3 : '#88559D',
  },
  dark_theme:{
    bg_color : '#000',
    primary_color : '#2ACA4B',
    // bg_color : '#293648',
    // primary_color : '#fff',
    ac_1 : '#2ACA4B',
    ac_2 : '#F9A5E4',
    ac_3 : '#88559D',
  }
}

module.exports = my_theme
