// buttons/day-button.js
const sheet = new CSSStyleSheet
sheet.replaceSync(get_theme())
/******************************************************************************
  DAY BUTTON COMPONENT
******************************************************************************/
// ----------------------------------------
// MODULE STATE & ID
var count = 0
const [cwd, dir] = [process.cwd(), __filename].map(x => new URL(x, 'file://').href)
const ID = dir.slice(cwd.length)
const STATE = { ids: {}, net: {} } // all state of component module
// ----------------------------------------
const default_opts = { }

module.exports = day_button

// Props - icon/img src
function day_button (protocol) {
  // ----------------------------------------
  // ID + JSON STATE
  // ----------------------------------------
  const name = `day_button`
  // ----------------------------------------
  // OPTS
  // ----------------------------------------
  const opts = {}
  // ----------------------------------------
  // PROTOCOL
  // ----------------------------------------
  const notify = protocol({ from: name }, listen)
  const PROTOCOL = {
    toggle_active,
    add_highlight,
    remove_highlight
  }
  // ----------------------------------------
  // TEMPLATE
  // ----------------------------------------
  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })
  shadow.adoptedStyleSheets = [sheet]
  shadow.innerHTML = `<div class="day_button"></div>`
  const day_button = shadow.querySelector(".day_button")
  // ----------------------------------------
  // ELEMENTS
  // ----------------------------------------
  // Toggle Icon
  day_button.onclick = (e) => {
    toggle_active()
    notify({
      head: { by: name, to: 'month_card', mid: 0 },
      type: 'toggle_day_button',
      data: el.id
    })
  }
  shadow.append(day_button)
  // ----------------------------------------
  // INIT
  // ----------------------------------------

  return el

  function listen (message) {
    const { head,  refs, type, data, meta } = message
    const { by, to, mid } = head
    PROTOCOL[type]()
  }
  function toggle_active () {
    day_button.classList.toggle('active')
  }
  function add_highlight () {
    day_button.classList.add('highlight')
  }
  function remove_highlight () {
    day_button.classList.remove('highlight')
  }
}
function get_theme () {
  return `
    .day_button {
      display: flex;
      justify-content: center;
      align-items: center;
      height: 20px;
      box-sizing: border-box;
      aspect-ratio: 1/1;
      cursor: pointer;
      border: 1px solid var(--primary_color);
      background-color: var(--bg_color);
      &.active {
        background-color: var(--ac-1) !important;
      }
      &.highlight {
        background-color: var(--ac-2)
      }
    }
  `
}
