// buttons/logo-button.js
const sheet = new CSSStyleSheet
sheet.replaceSync(get_theme())
/******************************************************************************
  LOGO BUTTON COMPONENT
******************************************************************************/
// ----------------------------------------
// MODULE STATE & ID
var count = 0
const [cwd, dir] = [process.cwd(), __filename].map(x => new URL(x, 'file://').href)
const ID = dir.slice(cwd.length)
const STATE = { ids: {}, net: {} } // all state of component module
// ----------------------------------------
const default_opts = { }

module.exports = logo_button

function logo_button (opts = default_opts) {
  // ----------------------------------------
  // ID + JSON STATE
  // ----------------------------------------
  const id = count++
  const path = require('path')
  const cwd = process.cwd()
  const prefix = path.relative(cwd, __dirname)
  // ----------------------------------------
  // OPTS
  // ----------------------------------------
  const { data } = opts
  // ----------------------------------------
  // TEMPLATE
  // ----------------------------------------
  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })
  shadow.innerHTML = `<div class="logo_button">
    <img src="${prefix}/logo.png" />
    <span> DAT ECOSYSTEM </span>
  </div>`
  shadow.adoptedStyleSheets = [sheet]
  // ----------------------------------------
  // ELEMENTS
  // ----------------------------------------
  // ----------------------------------------
  // INIT
  // ----------------------------------------

  return el
}
function get_theme () {
  return `
    .logo_button {
      width: 100%;
      height: 40px;
      box-sizing: border-box;
      padding: 10px;
      display: flex;
      justify-content: center;
      align-items: center;
      gap: 10px;
      background-color: var(--primary_color);
      color: var(--bg_color);
      font-size: 0.875em;
      letter-spacing: 0.25rem;
    }
  `
}
