// buttons/select-button.js
const scrollbar = require('scrollbar')
/******************************************************************************
  SELECT BUTTON COMPONENT
******************************************************************************/
// ----------------------------------------
// MODULE STATE & ID
var count = 0
const [cwd, dir] = [process.cwd(), __filename].map(x => new URL(x, 'file://').href)
const ID = dir.slice(cwd.length)
const STATE = { ids: {}, net: {} } // all state of component module
// ----------------------------------------
const sheet = new CSSStyleSheet
sheet.replaceSync(get_theme())
const default_opts = { }
const shopts = { mode: 'closed' }
// ----------------------------------------
module.exports = select_button
// ----------------------------------------
function select_button (opts = default_opts, protocol) {
  // ----------------------------------------
  // RESOURCE POOL (can't be serialized)
  // ----------------------------------------
  const ro = new ResizeObserver(entries => {
    console.log('ResizeObserver:terminal:resize')
    const scroll_channel = state.net[state.aka.scrollbar]
    scroll_channel.send({
      head: [id, scroll_channel.send.id, scroll_channel.mid++],
      refs: { },
      type: 'handle_scroll',
    })
  })
  // ----------------------------------------
  // ID + JSON STATE
  // ----------------------------------------
  const id = `${ID}:${count++}` // assigns their own name
  const status = {}
  const state = STATE.ids[id] = { id, status, wait: {}, net: {}, aka: {} } // all state of component instance
  const cache = resources({})
  let active_option = ''
  let active_state = true
  // ----------------------------------------
  // OPTS
  // ----------------------------------------
  const { data } = opts
  // Assigning all the icons
  const { img_src } = data
  const {
    icon_arrow_down,
    icon_arrow_up
  } = img_src
  // ----------------------------------------
  // PROTOCOL
  // ----------------------------------------
  const on = {}
  const channel = use_protocol('up')({ protocol, state, on })
  // ----------------------------------------
  // TEMPLATE
  // ----------------------------------------
  const el = document.createElement('div')
  const shadow = el.attachShadow(shopts)
  shadow.adoptedStyleSheets = [sheet]
  shadow.innerHTML = `<div tabindex="0" class="select_button_wrapper bottom">
    <div class="option_wrapper">
      <div class="option_scrollbar_wrapper">
        <div class="option active">ALL</div>
        ${opts.choices.map(choice => `<div class="option">${choice}</div>`).join('')}
      </div>
    </div>
    <div class="button_wrapper">
      <span class="button_name">${opts.name}: </span>
      <span class="selected_option">${'ALL'}</span>
      <span class="arrow_icon">
        ${icon_arrow_up}
      </span>
    </div>
  </div>`
  const select_button_wrapper = shadow.querySelector('.select_button_wrapper')
  const select_toggle_btn = shadow.querySelector('.button_wrapper')
  const options = shadow.querySelectorAll('.option')
  const selected_option = shadow.querySelector('.selected_option')
  const option_wrapper = shadow.querySelector('.option_wrapper')
  const option_scrollbar_wrapper = shadow.querySelector('.option_scrollbar_wrapper')
  // ----------------------------------------
  // ELEMENTS
  // ----------------------------------------
  select_toggle_btn.onclick = (e) => {
    select_button_wrapper.classList.toggle('active')
    shadow.querySelector('.arrow_icon').innerHTML = active_state ? icon_arrow_down : icon_arrow_up
    active_state = !active_state
  }
  select_button_wrapper.onblur = (e) => {
    setTimeout(() => {
      select_button_wrapper.classList.remove('active')
      shadow.querySelector('.arrow_icon').innerHTML = icon_arrow_up
      active_state = true
    }, 200);
  }
  { // scrollbar
    const on = { 'set_scroll': on_set_scroll, status: onstatus }
    const protocol = use_protocol('scrollbar')({ state, on })
    opts.data.img_src.icon_arrow_start = opts.data.img_src.icon_arrow_up
    opts.data.img_src.icon_arrow_end = opts.data.img_src.icon_arrow_down
    const opts1 = { data }
    const element = shadowfy()(scrollbar(opts1, protocol))
    option_wrapper.append(element)
    option_scrollbar_wrapper.onscroll = on_scroll
  }
  // select_toggle_btn.addEventListener('click', function() {
  //   shadow.querySelector('.select_button_wrapper').classList.toggle('active')
  // })
  // Use event delegation
  // document.addEventListener('click', (e) => {
  //   console.log(e.target.className)
  // })
  // Select all .option divs
  // Attach click event listener to each .option div
  options.forEach((option) => {
    option.addEventListener('click', () => {
      shadow.querySelector('.arrow_icon').innerHTML = active_state ? icon_arrow_down : icon_arrow_up
      active_state = !active_state

      if (active_option) active_option.classList.remove('active')
      if (active_option === option) {
        selected_option.innerHTML = 'ALL'
        active_option = options[0]
        options[0].classList.add('active')
      }
      else {
        option.classList.add('active')
        selected_option.innerHTML = option.innerHTML
        active_option = option
      }
      select_button_wrapper.classList.remove('active')
      channel.send({
        head: [id, channel.send.id, channel.mid++],
        type: 'value',
        data: { filter: opts.name, value: selected_option.innerHTML }
      })
    })
  })
  // ----------------------------------------
  // INIT
  // ----------------------------------------
  active_option = options[0]
  watch_scrollbar()

  return el

  function watch_scrollbar () {
    const channel = state.net[state.aka.scrollbar]
    ro.observe(option_scrollbar_wrapper)
  }
  function on_scroll (message) {
    const channel = state.net[state.aka.scrollbar]
    channel.send({
      head: [id, channel.send.id, channel.mid++],
      refs: { },
      type: 'handle_scroll',
    })
  }
  function on_set_scroll (message) {
    console.log('set_scroll', message) 
    setScrollTop(message.data)
  }
  function onstatus (message) {
    const channel = state.net[state.aka.scrollbar]
    channel.send({
      head: [id, channel.send.id, channel.mid++],
      refs: { cause: message.head },
      type: 'update_size',
      data: {
        sh: option_scrollbar_wrapper.scrollHeight,
        ch: option_scrollbar_wrapper.clientHeight,
        st: option_scrollbar_wrapper.scrollTop
      }
    })
  }
  async function setScrollTop (value) {
    option_scrollbar_wrapper.scrollTop = value
  }
}
function get_theme () {
  return `
    .select_button_wrapper {
      box-sizing: border-box;
      position: relative;
      z-index: 100;
      width: 100%;
      min-width:150px;
      height: 30px;
      font-size: 0.875em;
      line-height: 1.5em;
      background-color: var(--bg_color);
    }
    .select_button_wrapper.bottom .option_wrapper {
      bottom: 30px;
      left: 0px;
    }
    .select_button_wrapper.top .option_wrapper {
      /* top: 40px; */
      left: 0px;
    }
    .select_button_wrapper.active .option_wrapper{ display: flex !important; }
    .select_button_wrapper.active .option_wrapper .button_wrapper{ border: 2px solid var(--ac-1); }
    .select_button_wrapper .option_wrapper {
      position: absolute;
      display: none;
      box-sizing: border-box;
      height: max-content;
      max-height: 400px;
      width: 100%;
      background-color: var(--bg_color);
      border: 1px solid var(--primary_color);
      word-break: break-all;
    }
    
    .select_button_wrapper .option_wrapper .option {
      box-sizing: border-box;
      display: flex;
      gap: 5px;
      align-items: center;
      padding: 10px 5px;
      cursor: pointer;
      word-break: break-all;
      background-color: var(--bg_color);
      overflow: hidden;
    }
    .select_button_wrapper .option_wrapper .option.active {
      background-color: var(--ac-1);
      color: var(--primary_color);
    }
    .select_button_wrapper .option_wrapper .option:hover {
      filter: brightness(0.8);
    }
    .select_button_wrapper .button_wrapper {
      box-sizing: border-box;
      display: flex;
      align-items: center;
      gap: 5px;
      padding: 5px 5px;
      cursor: pointer;
      height: 30px;
      background-color: var(--bg_color);
      border: 1px solid var(--primary_color);
    }
    .select_button_wrapper .button_wrapper .button_name { 
      display: flex;
      vertical-align: middle;
      font-weight: 700;
      line-height: 15px;
      letter-spacing: -1px;
    }
    .select_button_wrapper .button_wrapper .selected_option { 
      display: flex;
      vertical-align: middle;
      font-weight: 300;
      line-height: 15px;
      letter-spacing: -1px;
      word-break: normal;
    }
    .select_button_wrapper .button_wrapper .arrow_icon {
      display: flex;
      align-items: center;
      margin-left: auto;
    }
    .select_button_wrapper .button_wrapper .arrow_icon svg * {
      fill: var(--primary_color);
    }
    .option_scrollbar_wrapper{
      height: max-content;
      max-height: 200px;
      min-height: 100px;
      overflow-y: scroll;
      scrollbar-width: none;
      width: 100%;
      background-color: var(--bg_color);
      border: 1px solid var(--primary_color);
    }
    .select_button_wrapper .option_scrollbar_wrapper::-webkit-scrollbar {
      display: none;
    }
  `
}
// ----------------------------------------------------------------------------
function shadowfy (props = {}, sheets = []) {
  return element => {
    const el = Object.assign(document.createElement('div'), { ...props })
    const sh = el.attachShadow(shopts)
    sh.adoptedStyleSheets = sheets
    sh.append(element)
    return el
  }
}
function use_protocol (petname) {
  return ({ protocol, state, on = { } }) => {
    if (petname in state.aka) throw new Error('petname already initialized')
    const { id } = state
    const invalid = on[''] || (message => console.error('invalid type', message))
    if (protocol) return handshake(protocol(Object.assign(listen, { id })))
    else return handshake
    // ----------------------------------------
    // @TODO: how to disconnect channel
    // ----------------------------------------
    function handshake (send) {
      state.aka[petname] = send.id
      const channel = state.net[send.id] = { petname, mid: 0, send, on }
      return protocol ? channel : Object.assign(listen, { id })
    }
    function listen (message) {
      const [from] = message.head
      const by = state.aka[petname]
      if (from !== by) return invalid(message) // @TODO: maybe forward
      console.log(`[${id}]:${petname}>`, message)
      const { on } = state.net[by]
      const action = on[message.type] || invalid
      action(message)
    }
  }
}
// ----------------------------------------------------------------------------
function resources (pool) {
  var num = 0
  return factory => {
    const prefix = num++
    const get = name => {
      const id = prefix + name
      if (pool[id]) return pool[id]
      const type = factory[name]
      return pool[id] = type()
    }
    return Object.assign(get, factory)
  }
}
