// buttons/sm-text-button.js
// CSS Boiler Plat
const sheet = new CSSStyleSheet
const theme = get_theme()
sheet.replaceSync(theme)

module.exports = sm_text_button

function sm_text_button (props) {
  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode:'closed' })
  shadow.innerHTML = `
    <div class="sm_text_button"> 
      ${props.text}
    </div>
    <style>${get_theme()}</style>
  `
  let sm_text_button = shadow.querySelector('.sm_text_button')
  sm_text_button.onclick = toggle_class
  shadow.adoptedStyleSheets = [sheet]
  
  return el

  function toggle_class (e) {
    let selector = e.target.classList
    ;( selector.contains('active') ) ? selector.remove('active') : selector.add('active')
  }
}
function get_theme () {
  return`
    .sm_text_button {
      text-align: center;
      font-size: 0.875em;
      line-height: .5em;
      padding: 10px 5px;
      height: 30px;
      box-sizing: border-box;
      width: 100%;
      cursor: pointer;
      border: 1px solid var(--primary_color);
      background-color: var(--bg_color);
      color: var(--primary_color);
      &.active{
        background-color: var(--ac-1);
        color: var(--primary_color);
      }
    }
  `
}
