// buttons/sm_text_button.js
module.exports = sm_text_button


// CSS Boiler Plat
const sheet = new CSSStyleSheet
const theme = get_theme()
sheet.replaceSync(theme)





function sm_text_button (props) {

    const el = document.createElement('div')
    const shadow = el.attachShadow({mode:'closed'})
    const sm_text_button = document.createElement('div')
    sm_text_button.classList.add('sm_text_button')
    sm_text_button.innerHTML = props.text
    sm_text_button.onclick = (e) => toggle_class(e)

    const style = document.createElement('style')
    style.textContent = get_theme()

    shadow.append(sm_text_button, style)
    shadow.adoptedStyleSheets = [sheet]
    return el

}



function get_theme(){
    return`
        .sm_text_button{
            text-align:center;
            font-size: 0.875em;
            line-height: .5em;
            padding:10px 5px;
            height:30px;
            box-sizing:border-box;
            width: 100%;
            cursor:pointer;
            border: 1px solid var(--primary_color);
            background-color: var(--bg_color);
            color:var(--primary_color);
        }
        .sm_text_button.active{
            background-color: var(--ac-1);
            color: var(--primary_color);
        }
    `
}



function toggle_class(e){
    let selector = e.target.classList
    ;( selector.contains('active') ) ? selector.remove('active') : selector.add('active')
}
