// buttons/year_button.js
module.exports = year_button


// CSS Boiler Plat
const sheet = new CSSStyleSheet
const theme = get_theme()
sheet.replaceSync(theme)




function year_button (props, protocol) {
    const name = 'year_button'
    const notify = protocol({from: name}, listen)

    const {data, latest_date} = props
    const { img_src : {
        icon_arrow_up= `${prefix}/icon_arrow_up.svg`,
    }} = data

    const el = document.createElement('div')
    const shadow = el.attachShadow({mode:'closed'})

    const date = new Date(latest_date)
    shadow.innerHTML = `
        <div class="year_button">
            <div class="text_wrapper">${date.getFullYear()}</div>
            ${icon_arrow_up}
        </div>
        <style> ${get_theme} </style>
    `
    const year_button = shadow.querySelector('.year_button')
    year_button.onclick = (e) => toggle_class(e)

    const text_wrapper = shadow.querySelector('.text_wrapper')


    shadow.adoptedStyleSheets = [sheet]
    return el

    function toggle_class(e){
        year_button.classList.toggle('active')
    }
    function listen(message){
        const {head, type, data} = message
        const {by, to, id} = head
        if(data.month || data.year)
            text_wrapper.innerHTML = `<b>${data.month.slice(0,3)}</b>${data.month && data.year && '/'}${data.year}`
        else
            text_wrapper.innerHTML = 'Select date'
    }
}



function get_theme(){
    return`
        .year_button{
            display: flex;
            cursor:pointer;
            box-sizing: border-box;
            border: 1px solid var(--primary_color);
            background-color: var(--bg_color);
            color:var(--primary_color);
            align-items: center;
            justify-content: center;
            padding: 0 4px;
            height:30px;
            width: 100%;
        }
        .text_wrapper{
            text-align:center;
            font-size: 0.875em;
            line-height: .5em;
            padding: 11px 0;
            height:30px;
            box-sizing:border-box;
            width: 100px;
            letter-spacing: -1px;
        }
        .year_button.active svg{
            rotate: 90deg;
        }
    `
}
